{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "orig = pd.read_csv(\"/oak/stanford/groups/akundaje/projects/cad/data/clint_data/rasqual_results/revisions/smc/SMC_rasqual_results_FDR_0.05_083121.txt\", sep=\"\\t\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature_ID</th>\n",
       "      <th>rsID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>SNP_position</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Freq</th>\n",
       "      <th>HWE_Chisquare</th>\n",
       "      <th>Imp_quality</th>\n",
       "      <th>Log10_BH_Q</th>\n",
       "      <th>...</th>\n",
       "      <th>Distance to TSS</th>\n",
       "      <th>Nearest PromoterID</th>\n",
       "      <th>Entrez ID</th>\n",
       "      <th>Nearest Unigene</th>\n",
       "      <th>Nearest Refseq</th>\n",
       "      <th>Nearest Ensembl</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Alias</th>\n",
       "      <th>Gene Description</th>\n",
       "      <th>Gene Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1.1069225.1069626</td>\n",
       "      <td>rs74048003</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1067007</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2000</td>\n",
       "      <td>0.156250</td>\n",
       "      <td>0.999916</td>\n",
       "      <td>-4.920961</td>\n",
       "      <td>...</td>\n",
       "      <td>4881</td>\n",
       "      <td>NM_001205252</td>\n",
       "      <td>401934.0</td>\n",
       "      <td>Hs.568137</td>\n",
       "      <td>NM_001205252</td>\n",
       "      <td>ENSG00000237330</td>\n",
       "      <td>RNF223</td>\n",
       "      <td>-</td>\n",
       "      <td>ring finger protein 223</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1.108118484.108118885</td>\n",
       "      <td>rs665659</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108121219</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.7250</td>\n",
       "      <td>0.000393</td>\n",
       "      <td>0.999935</td>\n",
       "      <td>-4.118572</td>\n",
       "      <td>...</td>\n",
       "      <td>74124</td>\n",
       "      <td>NM_213651</td>\n",
       "      <td>29957.0</td>\n",
       "      <td>Hs.656870</td>\n",
       "      <td>NM_013386</td>\n",
       "      <td>ENSG00000085491</td>\n",
       "      <td>SLC25A24</td>\n",
       "      <td>APC1|SCAMC-1|SCAMC1</td>\n",
       "      <td>solute carrier family 25 member 24</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1.108721337.108721738</td>\n",
       "      <td>rs4970806</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108721447</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.6625</td>\n",
       "      <td>5.930399</td>\n",
       "      <td>0.999939</td>\n",
       "      <td>-9.266900</td>\n",
       "      <td>...</td>\n",
       "      <td>8604</td>\n",
       "      <td>NM_001144937</td>\n",
       "      <td>163479.0</td>\n",
       "      <td>Hs.258253</td>\n",
       "      <td>NM_173532</td>\n",
       "      <td>ENSG00000143107</td>\n",
       "      <td>FNDC7</td>\n",
       "      <td>-</td>\n",
       "      <td>fibronectin type III domain containing 7</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1.108841698.108842099</td>\n",
       "      <td>rs34272714</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108842308</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.1375</td>\n",
       "      <td>1.016593</td>\n",
       "      <td>0.999885</td>\n",
       "      <td>-4.721082</td>\n",
       "      <td>...</td>\n",
       "      <td>15205</td>\n",
       "      <td>NM_152763</td>\n",
       "      <td>254268.0</td>\n",
       "      <td>Hs.729441</td>\n",
       "      <td>NM_152763</td>\n",
       "      <td>ENSG00000162641</td>\n",
       "      <td>AKNAD1</td>\n",
       "      <td>C1orf62</td>\n",
       "      <td>AKNA domain containing 1</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1.109274794.109275195</td>\n",
       "      <td>rs611917</td>\n",
       "      <td>chr1</td>\n",
       "      <td>109272630</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2750</td>\n",
       "      <td>5.566376</td>\n",
       "      <td>0.999931</td>\n",
       "      <td>-3.006430</td>\n",
       "      <td>...</td>\n",
       "      <td>8149</td>\n",
       "      <td>NM_001350237</td>\n",
       "      <td>84722.0</td>\n",
       "      <td>Hs.405925</td>\n",
       "      <td>NM_032636</td>\n",
       "      <td>ENSG00000134222</td>\n",
       "      <td>PSRC1</td>\n",
       "      <td>DDA3|FP3214</td>\n",
       "      <td>proline and serine rich coiled-coil 1</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Feature_ID        rsID Chromosome  SNP_position Ref_allele  \\\n",
       "0      chr1.1069225.1069626  rs74048003       chr1       1067007          A   \n",
       "1  chr1.108118484.108118885    rs665659       chr1     108121219          C   \n",
       "2  chr1.108721337.108721738   rs4970806       chr1     108721447          C   \n",
       "3  chr1.108841698.108842099  rs34272714       chr1     108842308          G   \n",
       "4  chr1.109274794.109275195    rs611917       chr1     109272630          A   \n",
       "\n",
       "  Alt_allele    Freq  HWE_Chisquare  Imp_quality  Log10_BH_Q  ...  \\\n",
       "0          G  0.2000       0.156250     0.999916   -4.920961  ...   \n",
       "1          G  0.7250       0.000393     0.999935   -4.118572  ...   \n",
       "2          T  0.6625       5.930399     0.999939   -9.266900  ...   \n",
       "3          C  0.1375       1.016593     0.999885   -4.721082  ...   \n",
       "4          G  0.2750       5.566376     0.999931   -3.006430  ...   \n",
       "\n",
       "   Distance to TSS  Nearest PromoterID  Entrez ID  Nearest Unigene  \\\n",
       "0             4881        NM_001205252   401934.0        Hs.568137   \n",
       "1            74124           NM_213651    29957.0        Hs.656870   \n",
       "2             8604        NM_001144937   163479.0        Hs.258253   \n",
       "3            15205           NM_152763   254268.0        Hs.729441   \n",
       "4             8149        NM_001350237    84722.0        Hs.405925   \n",
       "\n",
       "   Nearest Refseq  Nearest Ensembl  Gene Name           Gene Alias  \\\n",
       "0    NM_001205252  ENSG00000237330     RNF223                    -   \n",
       "1       NM_013386  ENSG00000085491   SLC25A24  APC1|SCAMC-1|SCAMC1   \n",
       "2       NM_173532  ENSG00000143107      FNDC7                    -   \n",
       "3       NM_152763  ENSG00000162641     AKNAD1              C1orf62   \n",
       "4       NM_032636  ENSG00000134222      PSRC1          DDA3|FP3214   \n",
       "\n",
       "                           Gene Description       Gene Type  \n",
       "0                   ring finger protein 223  protein-coding  \n",
       "1        solute carrier family 25 member 24  protein-coding  \n",
       "2  fibronectin type III domain containing 7  protein-coding  \n",
       "3                  AKNA domain containing 1  protein-coding  \n",
       "4     proline and serine rich coiled-coil 1  protein-coding  \n",
       "\n",
       "[5 rows x 47 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orig.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Feature_ID', 'rsID', 'Chromosome', 'SNP_position', 'Ref_allele',\n",
       "       'Alt_allele', 'Freq', 'HWE_Chisquare', 'Imp_quality', 'Log10_BH_Q',\n",
       "       'Chisquare', 'Effect_size', 'Delta', 'Phi', 'Overdispersion',\n",
       "       'SNP_id_region', 'Num_feature_SNPs', 'Num_tested_SNPs',\n",
       "       'Num_iterations_null', 'Num_iterations_alt', 'Random_ties',\n",
       "       'Log_likelihood_null', 'Convergence_status', 'Sq_corr_fSNPs',\n",
       "       'Sq_corr_rSNP', 'P', 'q', 'keep', 'PeakID', 'Chr', 'Start', 'End',\n",
       "       'Strand', 'Peak Score', 'Focus Ratio/Region Size', 'Annotation',\n",
       "       'Detailed Annotation', 'Distance to TSS', 'Nearest PromoterID',\n",
       "       'Entrez ID', 'Nearest Unigene', 'Nearest Refseq', 'Nearest Ensembl',\n",
       "       'Gene Name', 'Gene Alias', 'Gene Description', 'Gene Type'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orig.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "preds = pd.read_csv(\"/mnt/lab_data2/anusri/variant-scorer/src/output/smc_caqtls/ATAC_sig_only_new/summary.mean.variant_scores_new.tsv\", sep=\"\\t\", header=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1984, 23)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "preds[\"jsd.sign.mean\"] = np.sign(preds[\"logfc.mean\"])* preds[\"jsd.mean\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>rsid</th>\n",
       "      <th>logfc.mean</th>\n",
       "      <th>abs_logfc.mean</th>\n",
       "      <th>abs_logfc.mean.pval</th>\n",
       "      <th>jsd.mean</th>\n",
       "      <th>jsd.mean.pval</th>\n",
       "      <th>...</th>\n",
       "      <th>max_percentile.mean.pval</th>\n",
       "      <th>logfc_x_max_percentile.mean</th>\n",
       "      <th>abs_logfc_x_max_percentile.mean</th>\n",
       "      <th>abs_logfc_x_max_percentile.mean.pval</th>\n",
       "      <th>jsd_x_max_percentile.mean</th>\n",
       "      <th>jsd_x_max_percentile.mean.pval</th>\n",
       "      <th>abs_logfc_x_jsd_x_max_percentile.mean</th>\n",
       "      <th>abs_logfc_x_jsd_x_max_percentile.mean.pval</th>\n",
       "      <th>logfc_x_jsd_x_max_percentile.mean</th>\n",
       "      <th>jsd.sign.mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1067007</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1.1069225.1069626,rs74048003,chr1,1067007,A...</td>\n",
       "      <td>0.022735</td>\n",
       "      <td>0.027306</td>\n",
       "      <td>0.525433</td>\n",
       "      <td>0.008771</td>\n",
       "      <td>0.578309</td>\n",
       "      <td>...</td>\n",
       "      <td>0.071857</td>\n",
       "      <td>0.015619</td>\n",
       "      <td>0.019011</td>\n",
       "      <td>0.247607</td>\n",
       "      <td>0.005870</td>\n",
       "      <td>0.175994</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.302285</td>\n",
       "      <td>0.000184</td>\n",
       "      <td>0.008771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108121219</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1.108118484.108118885,rs665659,chr1,1081212...</td>\n",
       "      <td>0.021728</td>\n",
       "      <td>0.026806</td>\n",
       "      <td>0.533548</td>\n",
       "      <td>0.009595</td>\n",
       "      <td>0.513887</td>\n",
       "      <td>...</td>\n",
       "      <td>0.097398</td>\n",
       "      <td>0.013720</td>\n",
       "      <td>0.016709</td>\n",
       "      <td>0.288160</td>\n",
       "      <td>0.005725</td>\n",
       "      <td>0.178482</td>\n",
       "      <td>0.000164</td>\n",
       "      <td>0.334105</td>\n",
       "      <td>0.000141</td>\n",
       "      <td>0.009595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108721447</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1.108721337.108721738,rs4970806,chr1,108721...</td>\n",
       "      <td>0.289341</td>\n",
       "      <td>0.289341</td>\n",
       "      <td>0.034743</td>\n",
       "      <td>0.031146</td>\n",
       "      <td>0.047851</td>\n",
       "      <td>...</td>\n",
       "      <td>0.138296</td>\n",
       "      <td>0.148480</td>\n",
       "      <td>0.148480</td>\n",
       "      <td>0.018484</td>\n",
       "      <td>0.015681</td>\n",
       "      <td>0.031848</td>\n",
       "      <td>0.004771</td>\n",
       "      <td>0.023252</td>\n",
       "      <td>0.004771</td>\n",
       "      <td>0.031146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108842308</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1.108841698.108842099,rs34272714,chr1,10884...</td>\n",
       "      <td>-0.293381</td>\n",
       "      <td>0.293381</td>\n",
       "      <td>0.031649</td>\n",
       "      <td>0.041953</td>\n",
       "      <td>0.023738</td>\n",
       "      <td>...</td>\n",
       "      <td>0.193577</td>\n",
       "      <td>-0.127705</td>\n",
       "      <td>0.127705</td>\n",
       "      <td>0.023055</td>\n",
       "      <td>0.018106</td>\n",
       "      <td>0.024531</td>\n",
       "      <td>0.005598</td>\n",
       "      <td>0.019474</td>\n",
       "      <td>-0.005598</td>\n",
       "      <td>-0.041953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109272630</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1.109274794.109275195,rs611917,chr1,1092726...</td>\n",
       "      <td>0.100375</td>\n",
       "      <td>0.100375</td>\n",
       "      <td>0.171206</td>\n",
       "      <td>0.016716</td>\n",
       "      <td>0.188809</td>\n",
       "      <td>...</td>\n",
       "      <td>0.139277</td>\n",
       "      <td>0.051331</td>\n",
       "      <td>0.051331</td>\n",
       "      <td>0.086704</td>\n",
       "      <td>0.008469</td>\n",
       "      <td>0.103948</td>\n",
       "      <td>0.000975</td>\n",
       "      <td>0.102504</td>\n",
       "      <td>0.000975</td>\n",
       "      <td>0.016716</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr        pos allele1 allele2  \\\n",
       "0  chr1    1067007       A       G   \n",
       "1  chr1  108121219       C       G   \n",
       "2  chr1  108721447       C       T   \n",
       "3  chr1  108842308       G       C   \n",
       "4  chr1  109272630       A       G   \n",
       "\n",
       "                                                rsid  logfc.mean  \\\n",
       "0  chr1.1069225.1069626,rs74048003,chr1,1067007,A...    0.022735   \n",
       "1  chr1.108118484.108118885,rs665659,chr1,1081212...    0.021728   \n",
       "2  chr1.108721337.108721738,rs4970806,chr1,108721...    0.289341   \n",
       "3  chr1.108841698.108842099,rs34272714,chr1,10884...   -0.293381   \n",
       "4  chr1.109274794.109275195,rs611917,chr1,1092726...    0.100375   \n",
       "\n",
       "   abs_logfc.mean  abs_logfc.mean.pval  jsd.mean  jsd.mean.pval  ...  \\\n",
       "0        0.027306             0.525433  0.008771       0.578309  ...   \n",
       "1        0.026806             0.533548  0.009595       0.513887  ...   \n",
       "2        0.289341             0.034743  0.031146       0.047851  ...   \n",
       "3        0.293381             0.031649  0.041953       0.023738  ...   \n",
       "4        0.100375             0.171206  0.016716       0.188809  ...   \n",
       "\n",
       "   max_percentile.mean.pval  logfc_x_max_percentile.mean  \\\n",
       "0                  0.071857                     0.015619   \n",
       "1                  0.097398                     0.013720   \n",
       "2                  0.138296                     0.148480   \n",
       "3                  0.193577                    -0.127705   \n",
       "4                  0.139277                     0.051331   \n",
       "\n",
       "   abs_logfc_x_max_percentile.mean  abs_logfc_x_max_percentile.mean.pval  \\\n",
       "0                         0.019011                              0.247607   \n",
       "1                         0.016709                              0.288160   \n",
       "2                         0.148480                              0.018484   \n",
       "3                         0.127705                              0.023055   \n",
       "4                         0.051331                              0.086704   \n",
       "\n",
       "   jsd_x_max_percentile.mean  jsd_x_max_percentile.mean.pval  \\\n",
       "0                   0.005870                        0.175994   \n",
       "1                   0.005725                        0.178482   \n",
       "2                   0.015681                        0.031848   \n",
       "3                   0.018106                        0.024531   \n",
       "4                   0.008469                        0.103948   \n",
       "\n",
       "   abs_logfc_x_jsd_x_max_percentile.mean  \\\n",
       "0                               0.000209   \n",
       "1                               0.000164   \n",
       "2                               0.004771   \n",
       "3                               0.005598   \n",
       "4                               0.000975   \n",
       "\n",
       "   abs_logfc_x_jsd_x_max_percentile.mean.pval  \\\n",
       "0                                    0.302285   \n",
       "1                                    0.334105   \n",
       "2                                    0.023252   \n",
       "3                                    0.019474   \n",
       "4                                    0.102504   \n",
       "\n",
       "   logfc_x_jsd_x_max_percentile.mean  jsd.sign.mean  \n",
       "0                           0.000184       0.008771  \n",
       "1                           0.000141       0.009595  \n",
       "2                           0.004771       0.031146  \n",
       "3                          -0.005598      -0.041953  \n",
       "4                           0.000975       0.016716  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "preds = preds[[\"chr\", \"pos\", \"allele1\", \"allele2\", \"logfc.mean\", \"jsd.sign.mean\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "orig = orig.merge(preds, left_on=['Chromosome', 'SNP_position', 'Ref_allele','Alt_allele'], right_on=[\"chr\", \"pos\", \"allele1\", \"allele2\"]).drop_duplicates()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Feature_ID', 'rsID', 'Chromosome', 'SNP_position', 'Ref_allele',\n",
       "       'Alt_allele', 'Freq', 'HWE_Chisquare', 'Imp_quality', 'Log10_BH_Q',\n",
       "       'Chisquare', 'Effect_size', 'Delta', 'Phi', 'Overdispersion',\n",
       "       'SNP_id_region', 'Num_feature_SNPs', 'Num_tested_SNPs',\n",
       "       'Num_iterations_null', 'Num_iterations_alt', 'Random_ties',\n",
       "       'Log_likelihood_null', 'Convergence_status', 'Sq_corr_fSNPs',\n",
       "       'Sq_corr_rSNP', 'P', 'q', 'keep', 'PeakID', 'Chr', 'Start', 'End',\n",
       "       'Strand', 'Peak Score', 'Focus Ratio/Region Size', 'Annotation',\n",
       "       'Detailed Annotation', 'Distance to TSS', 'Nearest PromoterID',\n",
       "       'Entrez ID', 'Nearest Unigene', 'Nearest Refseq', 'Nearest Ensembl',\n",
       "       'Gene Name', 'Gene Alias', 'Gene Description', 'Gene Type', 'chr',\n",
       "       'pos', 'allele1', 'allele2', 'logfc.mean', 'jsd.sign.mean'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orig.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "orig = orig[['Chromosome', 'SNP_position', 'Ref_allele', 'Alt_allele', 'Log10_BH_Q', 'Effect_size', 'logfc.mean', 'jsd.sign.mean']]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks_bed = pd.read_csv(\"/oak/stanford/groups/akundaje/projects/cad/outs/7_30_2023_human_all_models/peaks/original/SMC.peaks.bed\", sep=\"\\t\", header=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2</td>\n",
       "      <td>624982</td>\n",
       "      <td>SMC_m2PE_peak_1</td>\n",
       "      <td>1978</td>\n",
       "      <td>.</td>\n",
       "      <td>8.30810</td>\n",
       "      <td>200.48422</td>\n",
       "      <td>197.87073</td>\n",
       "      <td>181479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>639033</td>\n",
       "      <td>773662</td>\n",
       "      <td>SMC_m2PE_peak_2</td>\n",
       "      <td>140</td>\n",
       "      <td>.</td>\n",
       "      <td>2.44222</td>\n",
       "      <td>15.88614</td>\n",
       "      <td>14.09891</td>\n",
       "      <td>131441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778301</td>\n",
       "      <td>779293</td>\n",
       "      <td>SMC_m2PE_peak_3</td>\n",
       "      <td>13192</td>\n",
       "      <td>.</td>\n",
       "      <td>12.68989</td>\n",
       "      <td>1324.53931</td>\n",
       "      <td>1319.24121</td>\n",
       "      <td>428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>783770</td>\n",
       "      <td>822546</td>\n",
       "      <td>SMC_m2PE_peak_4</td>\n",
       "      <td>4017</td>\n",
       "      <td>.</td>\n",
       "      <td>10.95845</td>\n",
       "      <td>404.78445</td>\n",
       "      <td>401.77557</td>\n",
       "      <td>33595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>826698</td>\n",
       "      <td>827834</td>\n",
       "      <td>SMC_m2PE_peak_5</td>\n",
       "      <td>11889</td>\n",
       "      <td>.</td>\n",
       "      <td>14.98692</td>\n",
       "      <td>1193.70923</td>\n",
       "      <td>1188.91516</td>\n",
       "      <td>827</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1       2                3      4  5         6           7  \\\n",
       "0  chr1       2  624982  SMC_m2PE_peak_1   1978  .   8.30810   200.48422   \n",
       "1  chr1  639033  773662  SMC_m2PE_peak_2    140  .   2.44222    15.88614   \n",
       "2  chr1  778301  779293  SMC_m2PE_peak_3  13192  .  12.68989  1324.53931   \n",
       "3  chr1  783770  822546  SMC_m2PE_peak_4   4017  .  10.95845   404.78445   \n",
       "4  chr1  826698  827834  SMC_m2PE_peak_5  11889  .  14.98692  1193.70923   \n",
       "\n",
       "            8       9  \n",
       "0   197.87073  181479  \n",
       "1    14.09891  131441  \n",
       "2  1319.24121     428  \n",
       "3   401.77557   33595  \n",
       "4  1188.91516     827  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>pos0</th>\n",
       "      <th>SNP_position</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Log10_BH_Q</th>\n",
       "      <th>Effect_size</th>\n",
       "      <th>logfc.mean</th>\n",
       "      <th>jsd.sign.mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1067006</td>\n",
       "      <td>1067007</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-4.920961</td>\n",
       "      <td>0.200715</td>\n",
       "      <td>0.022735</td>\n",
       "      <td>0.008771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108121218</td>\n",
       "      <td>108121219</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>-4.118572</td>\n",
       "      <td>0.228169</td>\n",
       "      <td>0.021728</td>\n",
       "      <td>0.009595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108721446</td>\n",
       "      <td>108721447</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-9.266900</td>\n",
       "      <td>0.764114</td>\n",
       "      <td>0.289341</td>\n",
       "      <td>0.031146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108842307</td>\n",
       "      <td>108842308</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-4.721082</td>\n",
       "      <td>0.242912</td>\n",
       "      <td>-0.293381</td>\n",
       "      <td>-0.041953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109272629</td>\n",
       "      <td>109272630</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-3.006430</td>\n",
       "      <td>0.709092</td>\n",
       "      <td>0.100375</td>\n",
       "      <td>0.016716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2033</th>\n",
       "      <td>chr9</td>\n",
       "      <td>96159156</td>\n",
       "      <td>96159157</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-3.026381</td>\n",
       "      <td>0.806026</td>\n",
       "      <td>-0.104170</td>\n",
       "      <td>-0.024419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2034</th>\n",
       "      <td>chr9</td>\n",
       "      <td>96178224</td>\n",
       "      <td>96178225</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-4.596132</td>\n",
       "      <td>0.787369</td>\n",
       "      <td>0.016330</td>\n",
       "      <td>0.007556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2035</th>\n",
       "      <td>chr9</td>\n",
       "      <td>97227899</td>\n",
       "      <td>97227900</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>-2.961709</td>\n",
       "      <td>0.405945</td>\n",
       "      <td>-0.000896</td>\n",
       "      <td>-0.007822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2036</th>\n",
       "      <td>chr9</td>\n",
       "      <td>98873906</td>\n",
       "      <td>98873907</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-4.182089</td>\n",
       "      <td>0.608833</td>\n",
       "      <td>-0.040235</td>\n",
       "      <td>-0.008717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2037</th>\n",
       "      <td>chr9</td>\n",
       "      <td>99444663</td>\n",
       "      <td>99444664</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-4.616533</td>\n",
       "      <td>0.279315</td>\n",
       "      <td>0.060264</td>\n",
       "      <td>0.012065</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1984 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome       pos0  SNP_position Ref_allele Alt_allele  Log10_BH_Q  \\\n",
       "0          chr1    1067006       1067007          A          G   -4.920961   \n",
       "1          chr1  108121218     108121219          C          G   -4.118572   \n",
       "2          chr1  108721446     108721447          C          T   -9.266900   \n",
       "3          chr1  108842307     108842308          G          C   -4.721082   \n",
       "4          chr1  109272629     109272630          A          G   -3.006430   \n",
       "...         ...        ...           ...        ...        ...         ...   \n",
       "2033       chr9   96159156      96159157          C          T   -3.026381   \n",
       "2034       chr9   96178224      96178225          G          A   -4.596132   \n",
       "2035       chr9   97227899      97227900          A          T   -2.961709   \n",
       "2036       chr9   98873906      98873907          T          C   -4.182089   \n",
       "2037       chr9   99444663      99444664          G          A   -4.616533   \n",
       "\n",
       "      Effect_size  logfc.mean  jsd.sign.mean  \n",
       "0        0.200715    0.022735       0.008771  \n",
       "1        0.228169    0.021728       0.009595  \n",
       "2        0.764114    0.289341       0.031146  \n",
       "3        0.242912   -0.293381      -0.041953  \n",
       "4        0.709092    0.100375       0.016716  \n",
       "...           ...         ...            ...  \n",
       "2033     0.806026   -0.104170      -0.024419  \n",
       "2034     0.787369    0.016330       0.007556  \n",
       "2035     0.405945   -0.000896      -0.007822  \n",
       "2036     0.608833   -0.040235      -0.008717  \n",
       "2037     0.279315    0.060264       0.012065  \n",
       "\n",
       "[1984 rows x 9 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orig[\"pos0\"] = orig[\"SNP_position\"] - 1\n",
    "selected_beds = orig[['Chromosome', \"pos0\", 'SNP_position', 'Ref_allele', 'Alt_allele', 'Log10_BH_Q', 'Effect_size', 'logfc.mean', 'jsd.sign.mean']]\n",
    "\n",
    "selected_beds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks_bed[1] = peaks_bed[1] + peaks_bed[9] - 100\n",
    "peaks_bed[2] = peaks_bed[1] + 200\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>181381</td>\n",
       "      <td>181581</td>\n",
       "      <td>SMC_m2PE_peak_1</td>\n",
       "      <td>1978</td>\n",
       "      <td>.</td>\n",
       "      <td>8.30810</td>\n",
       "      <td>200.48422</td>\n",
       "      <td>197.87073</td>\n",
       "      <td>181479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>770374</td>\n",
       "      <td>770574</td>\n",
       "      <td>SMC_m2PE_peak_2</td>\n",
       "      <td>140</td>\n",
       "      <td>.</td>\n",
       "      <td>2.44222</td>\n",
       "      <td>15.88614</td>\n",
       "      <td>14.09891</td>\n",
       "      <td>131441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778629</td>\n",
       "      <td>778829</td>\n",
       "      <td>SMC_m2PE_peak_3</td>\n",
       "      <td>13192</td>\n",
       "      <td>.</td>\n",
       "      <td>12.68989</td>\n",
       "      <td>1324.53931</td>\n",
       "      <td>1319.24121</td>\n",
       "      <td>428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>817265</td>\n",
       "      <td>817465</td>\n",
       "      <td>SMC_m2PE_peak_4</td>\n",
       "      <td>4017</td>\n",
       "      <td>.</td>\n",
       "      <td>10.95845</td>\n",
       "      <td>404.78445</td>\n",
       "      <td>401.77557</td>\n",
       "      <td>33595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>827425</td>\n",
       "      <td>827625</td>\n",
       "      <td>SMC_m2PE_peak_5</td>\n",
       "      <td>11889</td>\n",
       "      <td>.</td>\n",
       "      <td>14.98692</td>\n",
       "      <td>1193.70923</td>\n",
       "      <td>1188.91516</td>\n",
       "      <td>827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122115</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56831818</td>\n",
       "      <td>56832018</td>\n",
       "      <td>SMC_m2PE_peak_122116</td>\n",
       "      <td>66</td>\n",
       "      <td>.</td>\n",
       "      <td>1.44943</td>\n",
       "      <td>7.93707</td>\n",
       "      <td>6.62112</td>\n",
       "      <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122116</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56836744</td>\n",
       "      <td>56836944</td>\n",
       "      <td>SMC_m2PE_peak_122117</td>\n",
       "      <td>163</td>\n",
       "      <td>.</td>\n",
       "      <td>1.74992</td>\n",
       "      <td>18.21555</td>\n",
       "      <td>16.36313</td>\n",
       "      <td>135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122117</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56850861</td>\n",
       "      <td>56851061</td>\n",
       "      <td>SMC_m2PE_peak_122118</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>2.47251</td>\n",
       "      <td>25.62169</td>\n",
       "      <td>23.66050</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122118</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56870843</td>\n",
       "      <td>56871043</td>\n",
       "      <td>SMC_m2PE_peak_122119</td>\n",
       "      <td>292</td>\n",
       "      <td>.</td>\n",
       "      <td>3.14725</td>\n",
       "      <td>31.30095</td>\n",
       "      <td>29.28865</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122119</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56873684</td>\n",
       "      <td>56873884</td>\n",
       "      <td>SMC_m2PE_peak_122120</td>\n",
       "      <td>39</td>\n",
       "      <td>.</td>\n",
       "      <td>1.68549</td>\n",
       "      <td>5.09318</td>\n",
       "      <td>3.92608</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122120 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2                     3      4  5         6  \\\n",
       "0       chr1    181381    181581       SMC_m2PE_peak_1   1978  .   8.30810   \n",
       "1       chr1    770374    770574       SMC_m2PE_peak_2    140  .   2.44222   \n",
       "2       chr1    778629    778829       SMC_m2PE_peak_3  13192  .  12.68989   \n",
       "3       chr1    817265    817465       SMC_m2PE_peak_4   4017  .  10.95845   \n",
       "4       chr1    827425    827625       SMC_m2PE_peak_5  11889  .  14.98692   \n",
       "...      ...       ...       ...                   ...    ... ..       ...   \n",
       "122115  chrY  56831818  56832018  SMC_m2PE_peak_122116     66  .   1.44943   \n",
       "122116  chrY  56836744  56836944  SMC_m2PE_peak_122117    163  .   1.74992   \n",
       "122117  chrY  56850861  56851061  SMC_m2PE_peak_122118    236  .   2.47251   \n",
       "122118  chrY  56870843  56871043  SMC_m2PE_peak_122119    292  .   3.14725   \n",
       "122119  chrY  56873684  56873884  SMC_m2PE_peak_122120     39  .   1.68549   \n",
       "\n",
       "                 7           8       9  \n",
       "0        200.48422   197.87073  181479  \n",
       "1         15.88614    14.09891  131441  \n",
       "2       1324.53931  1319.24121     428  \n",
       "3        404.78445   401.77557   33595  \n",
       "4       1193.70923  1188.91516     827  \n",
       "...            ...         ...     ...  \n",
       "122115     7.93707     6.62112     223  \n",
       "122116    18.21555    16.36313     135  \n",
       "122117    25.62169    23.66050     373  \n",
       "122118    31.30095    29.28865      73  \n",
       "122119     5.09318     3.92608      36  \n",
       "\n",
       "[122120 rows x 10 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "beds_var = pybedtools.BedTool.from_dataframe(selected_beds)\n",
    "peak_bed =  pybedtools.BedTool.from_dataframe(peaks_bed)\n",
    "sig_vars_in_bed = beds_var.intersect(peak_bed, c=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_vars_in_bed = sig_vars_in_bed.to_dataframe(names=['Chromosome', \"pos0\", 'SNP_position', 'Ref_allele', 'Alt_allele', 'Log10_BH_Q', 'Effect_size', 'logfc.mean', 'jsd.sign.mean', \"hits\"]).drop_duplicates()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>pos0</th>\n",
       "      <th>SNP_position</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Log10_BH_Q</th>\n",
       "      <th>Effect_size</th>\n",
       "      <th>logfc.mean</th>\n",
       "      <th>jsd.sign.mean</th>\n",
       "      <th>hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1067006</td>\n",
       "      <td>1067007</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-4.920961</td>\n",
       "      <td>0.200715</td>\n",
       "      <td>0.022735</td>\n",
       "      <td>0.008771</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108121218</td>\n",
       "      <td>108121219</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>-4.118572</td>\n",
       "      <td>0.228169</td>\n",
       "      <td>0.021728</td>\n",
       "      <td>0.009595</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108721446</td>\n",
       "      <td>108721447</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-9.266900</td>\n",
       "      <td>0.764114</td>\n",
       "      <td>0.289341</td>\n",
       "      <td>0.031146</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108842307</td>\n",
       "      <td>108842308</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-4.721082</td>\n",
       "      <td>0.242912</td>\n",
       "      <td>-0.293381</td>\n",
       "      <td>-0.041953</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109272629</td>\n",
       "      <td>109272630</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-3.006430</td>\n",
       "      <td>0.709092</td>\n",
       "      <td>0.100375</td>\n",
       "      <td>0.016716</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome       pos0  SNP_position Ref_allele Alt_allele  Log10_BH_Q  \\\n",
       "0       chr1    1067006       1067007          A          G   -4.920961   \n",
       "1       chr1  108121218     108121219          C          G   -4.118572   \n",
       "2       chr1  108721446     108721447          C          T   -9.266900   \n",
       "3       chr1  108842307     108842308          G          C   -4.721082   \n",
       "4       chr1  109272629     109272630          A          G   -3.006430   \n",
       "\n",
       "   Effect_size  logfc.mean  jsd.sign.mean  hits  \n",
       "0     0.200715    0.022735       0.008771     0  \n",
       "1     0.228169    0.021728       0.009595     0  \n",
       "2     0.764114    0.289341       0.031146     1  \n",
       "3     0.242912   -0.293381      -0.041953     0  \n",
       "4     0.709092    0.100375       0.016716     0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_vars_in_bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1984, 10)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_vars_in_bed.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>pos0</th>\n",
       "      <th>SNP_position</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Log10_BH_Q</th>\n",
       "      <th>Effect_size</th>\n",
       "      <th>logfc.mean</th>\n",
       "      <th>jsd.sign.mean</th>\n",
       "      <th>hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1067006</td>\n",
       "      <td>1067007</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-4.920961</td>\n",
       "      <td>0.200715</td>\n",
       "      <td>0.022735</td>\n",
       "      <td>0.008771</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108121218</td>\n",
       "      <td>108121219</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>-4.118572</td>\n",
       "      <td>0.228169</td>\n",
       "      <td>0.021728</td>\n",
       "      <td>0.009595</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108721446</td>\n",
       "      <td>108721447</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-9.266900</td>\n",
       "      <td>0.764114</td>\n",
       "      <td>0.289341</td>\n",
       "      <td>0.031146</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108842307</td>\n",
       "      <td>108842308</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-4.721082</td>\n",
       "      <td>0.242912</td>\n",
       "      <td>-0.293381</td>\n",
       "      <td>-0.041953</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109272629</td>\n",
       "      <td>109272630</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-3.006430</td>\n",
       "      <td>0.709092</td>\n",
       "      <td>0.100375</td>\n",
       "      <td>0.016716</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979</th>\n",
       "      <td>chr9</td>\n",
       "      <td>96159156</td>\n",
       "      <td>96159157</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-3.026381</td>\n",
       "      <td>0.806026</td>\n",
       "      <td>-0.104170</td>\n",
       "      <td>-0.024419</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1980</th>\n",
       "      <td>chr9</td>\n",
       "      <td>96178224</td>\n",
       "      <td>96178225</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-4.596132</td>\n",
       "      <td>0.787369</td>\n",
       "      <td>0.016330</td>\n",
       "      <td>0.007556</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1981</th>\n",
       "      <td>chr9</td>\n",
       "      <td>97227899</td>\n",
       "      <td>97227900</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>-2.961709</td>\n",
       "      <td>0.405945</td>\n",
       "      <td>-0.000896</td>\n",
       "      <td>-0.007822</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1982</th>\n",
       "      <td>chr9</td>\n",
       "      <td>98873906</td>\n",
       "      <td>98873907</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-4.182089</td>\n",
       "      <td>0.608833</td>\n",
       "      <td>-0.040235</td>\n",
       "      <td>-0.008717</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1983</th>\n",
       "      <td>chr9</td>\n",
       "      <td>99444663</td>\n",
       "      <td>99444664</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-4.616533</td>\n",
       "      <td>0.279315</td>\n",
       "      <td>0.060264</td>\n",
       "      <td>0.012065</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1984 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome       pos0  SNP_position Ref_allele Alt_allele  Log10_BH_Q  \\\n",
       "0          chr1    1067006       1067007          A          G   -4.920961   \n",
       "1          chr1  108121218     108121219          C          G   -4.118572   \n",
       "2          chr1  108721446     108721447          C          T   -9.266900   \n",
       "3          chr1  108842307     108842308          G          C   -4.721082   \n",
       "4          chr1  109272629     109272630          A          G   -3.006430   \n",
       "...         ...        ...           ...        ...        ...         ...   \n",
       "1979       chr9   96159156      96159157          C          T   -3.026381   \n",
       "1980       chr9   96178224      96178225          G          A   -4.596132   \n",
       "1981       chr9   97227899      97227900          A          T   -2.961709   \n",
       "1982       chr9   98873906      98873907          T          C   -4.182089   \n",
       "1983       chr9   99444663      99444664          G          A   -4.616533   \n",
       "\n",
       "      Effect_size  logfc.mean  jsd.sign.mean  hits  \n",
       "0        0.200715    0.022735       0.008771     0  \n",
       "1        0.228169    0.021728       0.009595     0  \n",
       "2        0.764114    0.289341       0.031146     1  \n",
       "3        0.242912   -0.293381      -0.041953     0  \n",
       "4        0.709092    0.100375       0.016716     0  \n",
       "...           ...         ...            ...   ...  \n",
       "1979     0.806026   -0.104170      -0.024419     0  \n",
       "1980     0.787369    0.016330       0.007556     0  \n",
       "1981     0.405945   -0.000896      -0.007822     0  \n",
       "1982     0.608833   -0.040235      -0.008717     0  \n",
       "1983     0.279315    0.060264       0.012065     0  \n",
       "\n",
       "[1984 rows x 10 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_vars_in_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_vars_in_bed[\"IsUsed\"] = sig_vars_in_bed[\"hits\"] > 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>pos0</th>\n",
       "      <th>SNP_position</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Log10_BH_Q</th>\n",
       "      <th>Effect_size</th>\n",
       "      <th>logfc.mean</th>\n",
       "      <th>jsd.sign.mean</th>\n",
       "      <th>hits</th>\n",
       "      <th>IsUsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1067006</td>\n",
       "      <td>1067007</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-4.920961</td>\n",
       "      <td>0.200715</td>\n",
       "      <td>0.022735</td>\n",
       "      <td>0.008771</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108121218</td>\n",
       "      <td>108121219</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>-4.118572</td>\n",
       "      <td>0.228169</td>\n",
       "      <td>0.021728</td>\n",
       "      <td>0.009595</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108721446</td>\n",
       "      <td>108721447</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-9.266900</td>\n",
       "      <td>0.764114</td>\n",
       "      <td>0.289341</td>\n",
       "      <td>0.031146</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>108842307</td>\n",
       "      <td>108842308</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-4.721082</td>\n",
       "      <td>0.242912</td>\n",
       "      <td>-0.293381</td>\n",
       "      <td>-0.041953</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109272629</td>\n",
       "      <td>109272630</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-3.006430</td>\n",
       "      <td>0.709092</td>\n",
       "      <td>0.100375</td>\n",
       "      <td>0.016716</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979</th>\n",
       "      <td>chr9</td>\n",
       "      <td>96159156</td>\n",
       "      <td>96159157</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-3.026381</td>\n",
       "      <td>0.806026</td>\n",
       "      <td>-0.104170</td>\n",
       "      <td>-0.024419</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1980</th>\n",
       "      <td>chr9</td>\n",
       "      <td>96178224</td>\n",
       "      <td>96178225</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-4.596132</td>\n",
       "      <td>0.787369</td>\n",
       "      <td>0.016330</td>\n",
       "      <td>0.007556</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1981</th>\n",
       "      <td>chr9</td>\n",
       "      <td>97227899</td>\n",
       "      <td>97227900</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>-2.961709</td>\n",
       "      <td>0.405945</td>\n",
       "      <td>-0.000896</td>\n",
       "      <td>-0.007822</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1982</th>\n",
       "      <td>chr9</td>\n",
       "      <td>98873906</td>\n",
       "      <td>98873907</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-4.182089</td>\n",
       "      <td>0.608833</td>\n",
       "      <td>-0.040235</td>\n",
       "      <td>-0.008717</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1983</th>\n",
       "      <td>chr9</td>\n",
       "      <td>99444663</td>\n",
       "      <td>99444664</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-4.616533</td>\n",
       "      <td>0.279315</td>\n",
       "      <td>0.060264</td>\n",
       "      <td>0.012065</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1984 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome       pos0  SNP_position Ref_allele Alt_allele  Log10_BH_Q  \\\n",
       "0          chr1    1067006       1067007          A          G   -4.920961   \n",
       "1          chr1  108121218     108121219          C          G   -4.118572   \n",
       "2          chr1  108721446     108721447          C          T   -9.266900   \n",
       "3          chr1  108842307     108842308          G          C   -4.721082   \n",
       "4          chr1  109272629     109272630          A          G   -3.006430   \n",
       "...         ...        ...           ...        ...        ...         ...   \n",
       "1979       chr9   96159156      96159157          C          T   -3.026381   \n",
       "1980       chr9   96178224      96178225          G          A   -4.596132   \n",
       "1981       chr9   97227899      97227900          A          T   -2.961709   \n",
       "1982       chr9   98873906      98873907          T          C   -4.182089   \n",
       "1983       chr9   99444663      99444664          G          A   -4.616533   \n",
       "\n",
       "      Effect_size  logfc.mean  jsd.sign.mean  hits  IsUsed  \n",
       "0        0.200715    0.022735       0.008771     0   False  \n",
       "1        0.228169    0.021728       0.009595     0   False  \n",
       "2        0.764114    0.289341       0.031146     1    True  \n",
       "3        0.242912   -0.293381      -0.041953     0   False  \n",
       "4        0.709092    0.100375       0.016716     0   False  \n",
       "...           ...         ...            ...   ...     ...  \n",
       "1979     0.806026   -0.104170      -0.024419     0   False  \n",
       "1980     0.787369    0.016330       0.007556     0   False  \n",
       "1981     0.405945   -0.000896      -0.007822     0   False  \n",
       "1982     0.608833   -0.040235      -0.008717     0   False  \n",
       "1983     0.279315    0.060264       0.012065     0   False  \n",
       "\n",
       "[1984 rows x 11 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_vars_in_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols=['Chromosome', 'SNP_position', 'Ref_allele', 'Alt_allele', 'Log10_BH_Q', 'Effect_size', 'logfc.mean', 'jsd.sign.mean', \"IsUsed\"]\n",
    "\n",
    "sig_vars_in_bed[cols].to_csv(\"/mnt/lab_data2/anusri/chrombpnet/src/evaluation/figure_notebooks/figure_6/final_figures_new/smc.caqtls.tsv\", sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
