{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(28309, 18)\n"
     ]
    }
   ],
   "source": [
    "gkm_svm = pd.read_csv(\"/mnt/lab_data3/anusri/histone_expts/all_qtl_analysis/gm12878_sequence_sets/test_set/deltasv,/41588_2015_BFng3331_MOESM26_ESM.csv\", header=0,sep=',')\n",
    "print(gkm_svm.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_hg18</th>\n",
       "      <th>pos_hg18</th>\n",
       "      <th>SNPname1</th>\n",
       "      <th>chrom_hg19</th>\n",
       "      <th>pos_hg19</th>\n",
       "      <th>SNPname2</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>gkm_SVM</th>\n",
       "      <th>kmer_SVM</th>\n",
       "      <th>abs_gkm_SVM</th>\n",
       "      <th>abs_kmer_SVM</th>\n",
       "      <th>CADD</th>\n",
       "      <th>GWAVA_region</th>\n",
       "      <th>GWAVA_tss</th>\n",
       "      <th>GWAVA_unmatched</th>\n",
       "      <th>GERP</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>846446</td>\n",
       "      <td>chr1.846446</td>\n",
       "      <td>chr1</td>\n",
       "      <td>856583</td>\n",
       "      <td>rs141671872</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2.653531</td>\n",
       "      <td>7.938784</td>\n",
       "      <td>2.653531</td>\n",
       "      <td>7.938784</td>\n",
       "      <td>-1.522533</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.75</td>\n",
       "      <td>-0.682</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>901458</td>\n",
       "      <td>chr1.901458</td>\n",
       "      <td>chr1</td>\n",
       "      <td>911595</td>\n",
       "      <td>rs7417106</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2.821422</td>\n",
       "      <td>6.027648</td>\n",
       "      <td>2.821422</td>\n",
       "      <td>6.027648</td>\n",
       "      <td>1.739152</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.89</td>\n",
       "      <td>2.150</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1176365</td>\n",
       "      <td>chr1.1176365</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1186502</td>\n",
       "      <td>rs6603785</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>-7.167236</td>\n",
       "      <td>-10.870018</td>\n",
       "      <td>7.167236</td>\n",
       "      <td>10.870018</td>\n",
       "      <td>-0.091378</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.451</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1217275</td>\n",
       "      <td>chr1.1217275</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1227412</td>\n",
       "      <td>rs112571439</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-3.957382</td>\n",
       "      <td>2.387491</td>\n",
       "      <td>3.957382</td>\n",
       "      <td>2.387491</td>\n",
       "      <td>1.127472</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.43</td>\n",
       "      <td>0.49</td>\n",
       "      <td>2.430</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1580438</td>\n",
       "      <td>chr1.1580438</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1590575</td>\n",
       "      <td>rs9661285</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-4.785596</td>\n",
       "      <td>-6.827102</td>\n",
       "      <td>4.785596</td>\n",
       "      <td>6.827102</td>\n",
       "      <td>0.665464</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.86</td>\n",
       "      <td>1.640</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom_hg18  pos_hg18      SNPname1 chrom_hg19  pos_hg19     SNPname2  \\\n",
       "0       chr1    846446   chr1.846446       chr1    856583  rs141671872   \n",
       "1       chr1    901458   chr1.901458       chr1    911595    rs7417106   \n",
       "2       chr1   1176365  chr1.1176365       chr1   1186502    rs6603785   \n",
       "3       chr1   1217275  chr1.1217275       chr1   1227412  rs112571439   \n",
       "4       chr1   1580438  chr1.1580438       chr1   1590575    rs9661285   \n",
       "\n",
       "  allele1 allele2   gkm_SVM   kmer_SVM  abs_gkm_SVM  abs_kmer_SVM      CADD  \\\n",
       "0       A       G  2.653531   7.938784     2.653531      7.938784 -1.522533   \n",
       "1       G       A  2.821422   6.027648     2.821422      6.027648  1.739152   \n",
       "2       T       A -7.167236 -10.870018     7.167236     10.870018 -0.091378   \n",
       "3       A       G -3.957382   2.387491     3.957382      2.387491  1.127472   \n",
       "4       A       G -4.785596  -6.827102     4.785596      6.827102  0.665464   \n",
       "\n",
       "   GWAVA_region  GWAVA_tss  GWAVA_unmatched   GERP  label  \n",
       "0          0.48       0.53             0.75 -0.682      1  \n",
       "1          0.29       0.36             0.89  2.150      1  \n",
       "2          0.42       0.49             0.66  0.451      1  \n",
       "3          0.34       0.43             0.49  2.430      1  \n",
       "4          0.38       0.35             0.86  1.640      1  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gkm_svm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "574"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(gkm_svm[\"label\"]==1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27735"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(gkm_svm[\"label\"]==-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "es = pd.read_csv(\"/mnt/lab_data2/anusri/variant-scorer/src/output/dsqtls_lcl/GSE31388_dsQtlTable.txt.gz\", sep=\"\\t\")\n",
    "enformer_preds = pd.read_csv(\"/mnt/lab_data2/anusri/chrombpnet/results/variant_data/dsqtls/process/fetch_enformer/enformer_predictions.tsv\", sep=\"\\t\", header=0)\n",
    "gkm_svm['rsid'] = gkm_svm[\"chrom_hg19\"].astype(str) + \"_\" + gkm_svm[\"pos_hg19\"].astype(str) + \"_\" + gkm_svm[\"allele1\"].astype(str) + \"_\" + gkm_svm[\"allele2\"].astype(str)  \n",
    "es[\"rsid\"] = es[\"Chr\"]+\".\"+es[\"SNP\"].astype(str)\n",
    "gkm_svm = gkm_svm.merge(es, left_on=[\"SNPname1\"], right_on='rsid', how=\"left\").drop_duplicates('SNPname1')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(28309, 28)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gkm_svm.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_hg18</th>\n",
       "      <th>pos_hg18</th>\n",
       "      <th>SNPname1</th>\n",
       "      <th>chrom_hg19</th>\n",
       "      <th>pos_hg19</th>\n",
       "      <th>SNPname2</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>gkm_SVM</th>\n",
       "      <th>kmer_SVM</th>\n",
       "      <th>...</th>\n",
       "      <th>rsid_x</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Estimate</th>\n",
       "      <th>Std. Error</th>\n",
       "      <th>t value</th>\n",
       "      <th>Pr(&gt;|t|)</th>\n",
       "      <th>rsid_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>846446</td>\n",
       "      <td>chr1.846446</td>\n",
       "      <td>chr1</td>\n",
       "      <td>856583</td>\n",
       "      <td>rs141671872</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2.653531</td>\n",
       "      <td>7.938784</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1_856583_A_G</td>\n",
       "      <td>chr1</td>\n",
       "      <td>846400.0</td>\n",
       "      <td>846500.0</td>\n",
       "      <td>846446.0</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>0.17900</td>\n",
       "      <td>5.197</td>\n",
       "      <td>2.013000e-06</td>\n",
       "      <td>chr1.846446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>901458</td>\n",
       "      <td>chr1.901458</td>\n",
       "      <td>chr1</td>\n",
       "      <td>911595</td>\n",
       "      <td>rs7417106</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2.821422</td>\n",
       "      <td>6.027648</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1_911595_G_A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>901400.0</td>\n",
       "      <td>901500.0</td>\n",
       "      <td>901458.0</td>\n",
       "      <td>-0.6285</td>\n",
       "      <td>0.08721</td>\n",
       "      <td>-7.207</td>\n",
       "      <td>6.010000e-10</td>\n",
       "      <td>chr1.901458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1176365</td>\n",
       "      <td>chr1.1176365</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1186502</td>\n",
       "      <td>rs6603785</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>-7.167236</td>\n",
       "      <td>-10.870018</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1_1186502_T_A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1176300.0</td>\n",
       "      <td>1176400.0</td>\n",
       "      <td>1176365.0</td>\n",
       "      <td>-0.7363</td>\n",
       "      <td>0.08125</td>\n",
       "      <td>-9.062</td>\n",
       "      <td>2.607000e-13</td>\n",
       "      <td>chr1.1176365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1217275</td>\n",
       "      <td>chr1.1217275</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1227412</td>\n",
       "      <td>rs112571439</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-3.957382</td>\n",
       "      <td>2.387491</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1_1227412_A_G</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1217200.0</td>\n",
       "      <td>1217300.0</td>\n",
       "      <td>1217275.0</td>\n",
       "      <td>-0.9236</td>\n",
       "      <td>0.13360</td>\n",
       "      <td>-6.912</td>\n",
       "      <td>2.041000e-09</td>\n",
       "      <td>chr1.1217275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1580438</td>\n",
       "      <td>chr1.1580438</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1590575</td>\n",
       "      <td>rs9661285</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-4.785596</td>\n",
       "      <td>-6.827102</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1_1590575_A_G</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1580400.0</td>\n",
       "      <td>1580500.0</td>\n",
       "      <td>1580438.0</td>\n",
       "      <td>-1.0890</td>\n",
       "      <td>0.15100</td>\n",
       "      <td>-7.211</td>\n",
       "      <td>5.902000e-10</td>\n",
       "      <td>chr1.1580438</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom_hg18  pos_hg18      SNPname1 chrom_hg19  pos_hg19     SNPname2  \\\n",
       "0       chr1    846446   chr1.846446       chr1    856583  rs141671872   \n",
       "1       chr1    901458   chr1.901458       chr1    911595    rs7417106   \n",
       "2       chr1   1176365  chr1.1176365       chr1   1186502    rs6603785   \n",
       "3       chr1   1217275  chr1.1217275       chr1   1227412  rs112571439   \n",
       "4       chr1   1580438  chr1.1580438       chr1   1590575    rs9661285   \n",
       "\n",
       "  allele1 allele2   gkm_SVM   kmer_SVM  ...            rsid_x   Chr  \\\n",
       "0       A       G  2.653531   7.938784  ...   chr1_856583_A_G  chr1   \n",
       "1       G       A  2.821422   6.027648  ...   chr1_911595_G_A  chr1   \n",
       "2       T       A -7.167236 -10.870018  ...  chr1_1186502_T_A  chr1   \n",
       "3       A       G -3.957382   2.387491  ...  chr1_1227412_A_G  chr1   \n",
       "4       A       G -4.785596  -6.827102  ...  chr1_1590575_A_G  chr1   \n",
       "\n",
       "       Start        End        SNP  Estimate  Std. Error  t value  \\\n",
       "0   846400.0   846500.0   846446.0    0.9301     0.17900    5.197   \n",
       "1   901400.0   901500.0   901458.0   -0.6285     0.08721   -7.207   \n",
       "2  1176300.0  1176400.0  1176365.0   -0.7363     0.08125   -9.062   \n",
       "3  1217200.0  1217300.0  1217275.0   -0.9236     0.13360   -6.912   \n",
       "4  1580400.0  1580500.0  1580438.0   -1.0890     0.15100   -7.211   \n",
       "\n",
       "       Pr(>|t|)        rsid_y  \n",
       "0  2.013000e-06   chr1.846446  \n",
       "1  6.010000e-10   chr1.901458  \n",
       "2  2.607000e-13  chr1.1176365  \n",
       "3  2.041000e-09  chr1.1217275  \n",
       "4  5.902000e-10  chr1.1580438  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gkm_svm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "gkm_svm = gkm_svm[[\"chrom_hg19\", \"pos_hg19\", \"allele1\", \"allele2\",  \"SNPname2\", \"gkm_SVM\", \"Estimate\", \"label\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_hg19</th>\n",
       "      <th>pos_hg19</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>SNPname2</th>\n",
       "      <th>gkm_SVM</th>\n",
       "      <th>Estimate</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>856583</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs141671872</td>\n",
       "      <td>2.653531</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom_hg19  pos_hg19 allele1 allele2     SNPname2   gkm_SVM  Estimate  label\n",
       "0       chr1    856583       A       G  rs141671872  2.653531    0.9301      1"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gkm_svm.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsid</th>\n",
       "      <th>DNASE_SAD</th>\n",
       "      <th>DNASE_SAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>842250</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs148658266</td>\n",
       "      <td>-0.000669</td>\n",
       "      <td>-0.000892</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr     pos ref alt         rsid  DNASE_SAD  DNASE_SAR\n",
       "0  chr1  842250   A   G  rs148658266  -0.000669  -0.000892"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enformer_preds.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_enf_1 = enformer_preds.merge(gkm_svm, left_on=[\"rsid\", \"chr\", \"pos\", \"ref\", \"alt\"], right_on=[\"SNPname2\", \"chrom_hg19\", \"pos_hg19\", \"allele2\", \"allele1\"])\n",
    "merged_enf_2 = enformer_preds.merge(gkm_svm, left_on=[\"rsid\", \"chr\", \"pos\", \"ref\", \"alt\"], right_on=[\"SNPname2\", \"chrom_hg19\", \"pos_hg19\", \"allele1\", \"allele2\"])\n",
    "merged_enf_2[\"DNASE_SAD\"] = merged_enf_2[\"DNASE_SAD\"]*-1\n",
    "merged_enf_2[\"DNASE_SAR\"] = merged_enf_2[\"DNASE_SAR\"]*-1\n",
    "merged_enf = pd.concat([merged_enf_1, merged_enf_2]).drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsid</th>\n",
       "      <th>DNASE_SAD</th>\n",
       "      <th>DNASE_SAR</th>\n",
       "      <th>chrom_hg19</th>\n",
       "      <th>pos_hg19</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>SNPname2</th>\n",
       "      <th>gkm_SVM</th>\n",
       "      <th>Estimate</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>842250</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs148658266</td>\n",
       "      <td>-0.000669</td>\n",
       "      <td>-0.000892</td>\n",
       "      <td>chr1</td>\n",
       "      <td>842250</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs148658266</td>\n",
       "      <td>-3.62651</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr     pos ref alt         rsid  DNASE_SAD  DNASE_SAR chrom_hg19  \\\n",
       "0  chr1  842250   A   G  rs148658266  -0.000669  -0.000892       chr1   \n",
       "\n",
       "   pos_hg19 allele1 allele2     SNPname2  gkm_SVM  Estimate  label  \n",
       "0    842250       G       A  rs148658266 -3.62651       NaN     -1  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_enf.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27373\n",
      "560\n",
      "26813\n"
     ]
    }
   ],
   "source": [
    "rsid_samples = merged_enf[\"rsid\"].unique()\n",
    "gkm_svm[\"IsUsed\"] = False\n",
    "gkm_svm.loc[gkm_svm[\"SNPname2\"].isin(rsid_samples),\"IsUsed\"] = True\n",
    "\n",
    "print(sum(gkm_svm[\"IsUsed\"]))\n",
    "print(sum(gkm_svm[\"label\"]==1 & gkm_svm[\"IsUsed\"]))\n",
    "print(sum((gkm_svm[\"label\"]==-1) & gkm_svm[\"IsUsed\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(28309, 11)\n"
     ]
    }
   ],
   "source": [
    "data = gkm_svm.merge(merged_enf[[\"chrom_hg19\", \"pos_hg19\", \"allele1\", \"allele2\", \"DNASE_SAD\", \"DNASE_SAR\"]] , on=[\"chrom_hg19\", \"pos_hg19\", \"allele1\", \"allele2\"], how=\"left\")\n",
    "print(data.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "936"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "sum(~data[\"IsUsed\"] & np.isnan(data[\"DNASE_SAD\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_hg19</th>\n",
       "      <th>pos_hg19</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>SNPname2</th>\n",
       "      <th>gkm_SVM</th>\n",
       "      <th>Estimate</th>\n",
       "      <th>label</th>\n",
       "      <th>IsUsed</th>\n",
       "      <th>DNASE_SAD</th>\n",
       "      <th>DNASE_SAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>856583</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs141671872</td>\n",
       "      <td>2.653531</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000342</td>\n",
       "      <td>0.000450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>911595</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs7417106</td>\n",
       "      <td>2.821422</td>\n",
       "      <td>-0.6285</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.001241</td>\n",
       "      <td>-0.001526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1186502</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>rs6603785</td>\n",
       "      <td>-7.167236</td>\n",
       "      <td>-0.7363</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.006530</td>\n",
       "      <td>-0.007840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1227412</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs112571439</td>\n",
       "      <td>-3.957382</td>\n",
       "      <td>-0.9236</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.001878</td>\n",
       "      <td>-0.002290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1590575</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs9661285</td>\n",
       "      <td>-4.785596</td>\n",
       "      <td>-1.0890</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.008736</td>\n",
       "      <td>-0.010230</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom_hg19  pos_hg19 allele1 allele2     SNPname2   gkm_SVM  Estimate  \\\n",
       "0       chr1    856583       A       G  rs141671872  2.653531    0.9301   \n",
       "1       chr1    911595       G       A    rs7417106  2.821422   -0.6285   \n",
       "2       chr1   1186502       T       A    rs6603785 -7.167236   -0.7363   \n",
       "3       chr1   1227412       A       G  rs112571439 -3.957382   -0.9236   \n",
       "4       chr1   1590575       A       G    rs9661285 -4.785596   -1.0890   \n",
       "\n",
       "   label  IsUsed  DNASE_SAD  DNASE_SAR  \n",
       "0      1    True   0.000342   0.000450  \n",
       "1      1    True  -0.001241  -0.001526  \n",
       "2      1    True  -0.006530  -0.007840  \n",
       "3      1    True  -0.001878  -0.002290  \n",
       "4      1    True  -0.008736  -0.010230  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_preds=\"/mnt/lab_data2/anusri/variant-scorer/src/output/dsqtls_lcl/ATAC/summary.mean.variant_scores_new_2.tsv\"\n",
    "dnase_preds=\"/mnt/lab_data2/anusri/variant-scorer/src/output/dsqtls_lcl/DNASE_NEW/summary.mean.variant_scores_new_3.tsv\"\n",
    "atac_preds_50m=\"/mnt/lab_data2/anusri/variant-scorer/src/output/dsqtls_lcl/ATAC_50M/summary.mean.variant_scores_new_2.tsv\"\n",
    "\n",
    "atac_preds_250m=\"/mnt/lab_data2/anusri/variant-scorer/src/output/dsqtls_lcl/ATAC_250M/summary.mean.variant_scores_new_2.tsv\"\n",
    "atac_preds_100m=\"/mnt/lab_data2/anusri/variant-scorer/src/output/dsqtls_lcl/ATAC_100M/summary.mean.variant_scores_new_2.tsv\"\n",
    "atac_preds_25m=\"/mnt/lab_data2/anusri/variant-scorer/src/output/dsqtls_lcl/ATAC_25M/summary.mean.variant_scores_new_2.tsv\"\n",
    "atac_preds_5m=\"/mnt/lab_data2/anusri/variant-scorer/src/output/dsqtls_lcl/ATAC_5M/summary.mean.variant_scores_new_2.tsv\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sel_columns_and_append(df_f, out_df, title):\n",
    "    \n",
    "    keys = [\"abs_logfc_x_jsd_x_max_percentile.mean\",\n",
    "        \"abs_logfc_x_jsd_x_max_percentile.mean.pval\",\n",
    "        \"abs_logfc_x_jsd.mean\" , \n",
    "        \"jsd_x_max_percentile.mean\", \n",
    "        \"abs_logfc_x_max_percentile.mean\", \n",
    "        \"jsd.mean\", \n",
    "        \"abs_logfc.mean\",\n",
    "        \"max_percentile.mean\"]\n",
    "    \n",
    "    df = pd.read_csv(df_f, sep='\\t', header=0)\n",
    "\n",
    "    for key in keys:\n",
    "        \n",
    "        out_df.loc[:,title+\".\"+key] = df[key]\n",
    "    \n",
    "    \n",
    "    out_df[title+\".\"+\"logfc.mean\"] = df[\"logfc.mean\"]*-1\n",
    "    out_df[title+\".\"+\"sign.jsd.mean\"] = np.sign(df[\"logfc.mean\"])*df[\"jsd.mean\"]*-1\n",
    "    \n",
    "    return out_df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data[['chrom_hg19', 'pos_hg19', 'allele1', 'allele2', 'SNPname2', 'Estimate', 'label',  'IsUsed', 'gkm_SVM', 'DNASE_SAD', 'DNASE_SAR']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = get_sel_columns_and_append(atac_preds, data, \"ENCSR637XSC\")\n",
    "data = get_sel_columns_and_append(dnase_preds, data, \"ENCSR000EMT\")\n",
    "data = get_sel_columns_and_append(atac_preds_50m, data, \"ENCSR637XSC.sub.50M\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sel_columns_and_append(df_f, out_df, title):\n",
    "    \n",
    "    keys = [\"abs_logfc_x_jsd_x_max_percentile.mean\",\n",
    "        \"abs_logfc_x_jsd_x_max_percentile.mean.pval\"]\n",
    "    \n",
    "    df = pd.read_csv(df_f, sep='\\t', header=0)\n",
    "\n",
    "    for key in keys:\n",
    "        \n",
    "        out_df.loc[:,title+\".\"+key] = df[key]\n",
    "    \n",
    "    \n",
    "    out_df[title+\".\"+\"logfc.mean\"] = df[\"logfc.mean\"]*-1\n",
    "    out_df[title+\".\"+\"sign.jsd.mean\"] = np.sign(df[\"logfc.mean\"])*df[\"jsd.mean\"]*-1\n",
    "        \n",
    "    return out_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = get_sel_columns_and_append(atac_preds_250m, data, \"ENCSR637XSC.sub.250M\")\n",
    "data = get_sel_columns_and_append(atac_preds_100m, data, \"ENCSR637XSC.sub.100M\")\n",
    "data = get_sel_columns_and_append(atac_preds_25m, data, \"ENCSR637XSC.sub.25M\")\n",
    "data = get_sel_columns_and_append(atac_preds_5m, data, \"ENCSR637XSC.sub.5M\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(28309, 57)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_hg19</th>\n",
       "      <th>pos_hg19</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>SNPname2</th>\n",
       "      <th>Estimate</th>\n",
       "      <th>label</th>\n",
       "      <th>IsUsed</th>\n",
       "      <th>gkm_SVM</th>\n",
       "      <th>DNASE_SAD</th>\n",
       "      <th>...</th>\n",
       "      <th>ENCSR637XSC.sub.100M.logfc.mean</th>\n",
       "      <th>ENCSR637XSC.sub.100M.sign.jsd.mean</th>\n",
       "      <th>ENCSR637XSC.sub.25M.abs_logfc_x_jsd_x_max_percentile.mean</th>\n",
       "      <th>ENCSR637XSC.sub.25M.abs_logfc_x_jsd_x_max_percentile.mean.pval</th>\n",
       "      <th>ENCSR637XSC.sub.25M.logfc.mean</th>\n",
       "      <th>ENCSR637XSC.sub.25M.sign.jsd.mean</th>\n",
       "      <th>ENCSR637XSC.sub.5M.abs_logfc_x_jsd_x_max_percentile.mean</th>\n",
       "      <th>ENCSR637XSC.sub.5M.abs_logfc_x_jsd_x_max_percentile.mean.pval</th>\n",
       "      <th>ENCSR637XSC.sub.5M.logfc.mean</th>\n",
       "      <th>ENCSR637XSC.sub.5M.sign.jsd.mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>856583</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs141671872</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>2.653531</td>\n",
       "      <td>0.000342</td>\n",
       "      <td>...</td>\n",
       "      <td>0.61987</td>\n",
       "      <td>0.091883</td>\n",
       "      <td>0.00484</td>\n",
       "      <td>0.018692</td>\n",
       "      <td>0.291935</td>\n",
       "      <td>0.067401</td>\n",
       "      <td>0.00136</td>\n",
       "      <td>0.065265</td>\n",
       "      <td>0.065098</td>\n",
       "      <td>0.049902</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom_hg19  pos_hg19 allele1 allele2     SNPname2  Estimate  label  IsUsed  \\\n",
       "0       chr1    856583       A       G  rs141671872    0.9301      1    True   \n",
       "\n",
       "    gkm_SVM  DNASE_SAD  ...  ENCSR637XSC.sub.100M.logfc.mean  \\\n",
       "0  2.653531   0.000342  ...                          0.61987   \n",
       "\n",
       "   ENCSR637XSC.sub.100M.sign.jsd.mean  \\\n",
       "0                            0.091883   \n",
       "\n",
       "   ENCSR637XSC.sub.25M.abs_logfc_x_jsd_x_max_percentile.mean  \\\n",
       "0                                            0.00484           \n",
       "\n",
       "   ENCSR637XSC.sub.25M.abs_logfc_x_jsd_x_max_percentile.mean.pval  \\\n",
       "0                                           0.018692                \n",
       "\n",
       "   ENCSR637XSC.sub.25M.logfc.mean  ENCSR637XSC.sub.25M.sign.jsd.mean  \\\n",
       "0                        0.291935                           0.067401   \n",
       "\n",
       "   ENCSR637XSC.sub.5M.abs_logfc_x_jsd_x_max_percentile.mean  \\\n",
       "0                                            0.00136          \n",
       "\n",
       "   ENCSR637XSC.sub.5M.abs_logfc_x_jsd_x_max_percentile.mean.pval  \\\n",
       "0                                           0.065265               \n",
       "\n",
       "   ENCSR637XSC.sub.5M.logfc.mean  ENCSR637XSC.sub.5M.sign.jsd.mean  \n",
       "0                       0.065098                          0.049902  \n",
       "\n",
       "[1 rows x 57 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chrom_hg19', 'pos_hg19', 'allele1', 'allele2', 'SNPname2', 'Estimate',\n",
       "       'label', 'IsUsed', 'gkm_SVM', 'DNASE_SAD', 'DNASE_SAR',\n",
       "       'ENCSR637XSC.abs_logfc_x_jsd_x_max_percentile.mean',\n",
       "       'ENCSR637XSC.abs_logfc_x_jsd_x_max_percentile.mean.pval',\n",
       "       'ENCSR637XSC.abs_logfc_x_jsd.mean',\n",
       "       'ENCSR637XSC.jsd_x_max_percentile.mean',\n",
       "       'ENCSR637XSC.abs_logfc_x_max_percentile.mean', 'ENCSR637XSC.jsd.mean',\n",
       "       'ENCSR637XSC.abs_logfc.mean', 'ENCSR637XSC.max_percentile.mean',\n",
       "       'ENCSR637XSC.logfc.mean', 'ENCSR637XSC.sign.jsd.mean',\n",
       "       'ENCSR000EMT.abs_logfc_x_jsd_x_max_percentile.mean',\n",
       "       'ENCSR000EMT.abs_logfc_x_jsd_x_max_percentile.mean.pval',\n",
       "       'ENCSR000EMT.abs_logfc_x_jsd.mean',\n",
       "       'ENCSR000EMT.jsd_x_max_percentile.mean',\n",
       "       'ENCSR000EMT.abs_logfc_x_max_percentile.mean', 'ENCSR000EMT.jsd.mean',\n",
       "       'ENCSR000EMT.abs_logfc.mean', 'ENCSR000EMT.max_percentile.mean',\n",
       "       'ENCSR000EMT.logfc.mean', 'ENCSR000EMT.sign.jsd.mean',\n",
       "       'ENCSR637XSC.sub.50M.abs_logfc_x_jsd_x_max_percentile.mean',\n",
       "       'ENCSR637XSC.sub.50M.abs_logfc_x_jsd_x_max_percentile.mean.pval',\n",
       "       'ENCSR637XSC.sub.50M.abs_logfc_x_jsd.mean',\n",
       "       'ENCSR637XSC.sub.50M.jsd_x_max_percentile.mean',\n",
       "       'ENCSR637XSC.sub.50M.abs_logfc_x_max_percentile.mean',\n",
       "       'ENCSR637XSC.sub.50M.jsd.mean', 'ENCSR637XSC.sub.50M.abs_logfc.mean',\n",
       "       'ENCSR637XSC.sub.50M.max_percentile.mean',\n",
       "       'ENCSR637XSC.sub.50M.logfc.mean', 'ENCSR637XSC.sub.50M.sign.jsd.mean',\n",
       "       'ENCSR637XSC.sub.250M.abs_logfc_x_jsd_x_max_percentile.mean',\n",
       "       'ENCSR637XSC.sub.250M.abs_logfc_x_jsd_x_max_percentile.mean.pval',\n",
       "       'ENCSR637XSC.sub.250M.logfc.mean', 'ENCSR637XSC.sub.250M.sign.jsd.mean',\n",
       "       'ENCSR637XSC.sub.100M.abs_logfc_x_jsd_x_max_percentile.mean',\n",
       "       'ENCSR637XSC.sub.100M.abs_logfc_x_jsd_x_max_percentile.mean.pval',\n",
       "       'ENCSR637XSC.sub.100M.logfc.mean', 'ENCSR637XSC.sub.100M.sign.jsd.mean',\n",
       "       'ENCSR637XSC.sub.25M.abs_logfc_x_jsd_x_max_percentile.mean',\n",
       "       'ENCSR637XSC.sub.25M.abs_logfc_x_jsd_x_max_percentile.mean.pval',\n",
       "       'ENCSR637XSC.sub.25M.logfc.mean', 'ENCSR637XSC.sub.25M.sign.jsd.mean',\n",
       "       'ENCSR637XSC.sub.5M.abs_logfc_x_jsd_x_max_percentile.mean',\n",
       "       'ENCSR637XSC.sub.5M.abs_logfc_x_jsd_x_max_percentile.mean.pval',\n",
       "       'ENCSR637XSC.sub.5M.logfc.mean', 'ENCSR637XSC.sub.5M.sign.jsd.mean'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle as pkl\n",
    "import os\n",
    "enformer_recompute = pkl.load(open(os.path.join(\"/mnt/lab_data2/anusri/variant-scorer/src/output/dsqtls_lcl/enformer_preds_small_window/predictions_at_snp.pkl\"),'rb'))\n",
    "mean_change=[]\n",
    "for i in range(data.shape[0]):\n",
    "    mean_change.append(np.log2( (enformer_recompute[\"ref_logcount_preds\"][i,440+4:456-4].sum()+1) / (enformer_recompute[\"alt_logcount_preds\"][i,440+4:456-4].sum()+1)))\n",
    "    \n",
    "mean_change = np.array(mean_change)\n",
    "                       \n",
    "      \n",
    "data[\"pred.enformer.encsr000emt.varscore.local_logfc\"] = mean_change"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_rename = {\"chrom_hg19\": \"var.chrom\",\n",
    "                            \"SNPname2\": \"var.rsid\",\n",
    "                            \"DNASE_SAD\": \"pred.enformer.encsr000emt.varscore.global_sad\",\n",
    "                            \"DNASE_SAR\": \"pred.enformer.encsr000emt.varscore.global_sar\",\n",
    "                            \"pos_hg19\": \"var.pos_hg19\",\n",
    "                            \"allele1\": \"var.allele1\",\n",
    "                            \"allele2\": \"var.allele2\",\n",
    "                            \"Estimate\": \"obs.estimate\",\n",
    "                            \"label\" : \"obs.label\",\n",
    "                            \"IsUsed\": \"var.isused\",\n",
    "                            \"gkm_SVM\": \"pred.gkmsvm.encsr000emt.varscore.deltasvm\",\n",
    "                            \"ENCSR637XSC.abs_logfc_x_jsd_x_max_percentile.mean\": \"pred.chrombpnet.encsr637xsc.varscore.ips\",\n",
    "                            \"ENCSR000EMT.abs_logfc_x_jsd_x_max_percentile.mean\": \"pred.chrombpnet.encsr000emt.varscore.ips\",\n",
    "                            \"ENCSR637XSC.sub.50M.abs_logfc_x_jsd_x_max_percentile.mean\": \"pred.chrombpnet.encsr637xsc_50m.varscore.ips\",\n",
    "\n",
    "                            \"ENCSR637XSC.abs_logfc_x_jsd_x_max_percentile.mean.pval\": \"pred.chrombpnet.encsr637xsc.varscore.pval\",\n",
    "                            \"ENCSR000EMT.abs_logfc_x_jsd_x_max_percentile.mean.pval\": \"pred.chrombpnet.encsr000emt.varscore.pval\",\n",
    "                            \"ENCSR637XSC.sub.50M.abs_logfc_x_jsd_x_max_percentile.mean.pval\": \"pred.chrombpnet.encsr637xsc_50m.varscore.pval\",\n",
    "\n",
    "                            \"ENCSR637XSC.logfc.mean\": \"pred.chrombpnet.encsr637xsc.varscore.logfc\",\n",
    "                            \"ENCSR000EMT.logfc.mean\": \"pred.chrombpnet.encsr000emt.varscore.logfc\",\n",
    "                            \"ENCSR637XSC.sub.50M.logfc.mean\": \"pred.chrombpnet.encsr637xsc_50m.varscore.logfc\",\n",
    "\n",
    "                            \"ENCSR637XSC.sign.jsd.mean\": \"pred.chrombpnet.encsr637xsc.varscore.jsd\",\n",
    "                            \"ENCSR000EMT.sign.jsd.mean\": \"pred.chrombpnet.encsr000emt.varscore.jsd\",\n",
    "                            \"ENCSR637XSC.sub.50M.sign.jsd.mean\": \"pred.chrombpnet.encsr637xsc_50m.varscore.jsd\",\n",
    "\n",
    "                            \"ENCSR637XSC.max_percentile.mean\": \"pred.chrombpnet.encsr637xsc.varscore.aaq\",\n",
    "                            \"ENCSR000EMT.max_percentile.mean\": \"pred.chrombpnet.encsr000emt.varscore.aaq\",\n",
    "                            \"ENCSR637XSC.sub.50M.max_percentile.mean\": \"pred.chrombpnet.encsr637xsc_50m.varscore.aaq\",\n",
    "               \n",
    "                            \"ENCSR637XSC.abs_logfc_x_jsd.mean\": \"pred.chrombpnet.encsr637xsc.varscore.ies\",\n",
    "                            \"ENCSR000EMT.abs_logfc_x_jsd.mean\": \"pred.chrombpnet.encsr000emt.varscore.ies\",\n",
    "                            \"ENCSR637XSC.sub.50M.abs_logfc_x_jsd.mean\": \"pred.chrombpnet.encsr637xsc_50m.varscore.ies\",\n",
    "\n",
    "                            \"ENCSR637XSC.jsd_x_max_percentile.mean\": \"pred.chrombpnet.encsr637xsc.varscore.jsd_x_aaq\",\n",
    "                            \"ENCSR000EMT.jsd_x_max_percentile.mean\": \"pred.chrombpnet.encsr000emt.varscore.jsd_x_aaq\",\n",
    "                            \"ENCSR637XSC.sub.50M.jsd_x_max_percentile.mean\": \"pred.chrombpnet.encsr637xsc_50m.varscore.jsd_x_aaq\",\n",
    "               \n",
    "                            \"ENCSR637XSC.abs_logfc_x_max_percentile.mean\": \"pred.chrombpnet.encsr637xsc.varscore.logfc_x_aaq\",\n",
    "                            \"ENCSR000EMT.abs_logfc_x_max_percentile.mean\": \"pred.chrombpnet.encsr000emt.varscore.logfc_x_aaq\",\n",
    "                            \"ENCSR637XSC.sub.50M.abs_logfc_x_max_percentile.mean\": \"pred.chrombpnet.encsr637xsc_50m.varscore.logfc_x_aaq\"}\n",
    "\n",
    "\n",
    "keyds = list(dict_rename.keys())\n",
    "for key in keyds:\n",
    "    if \"50M\" in key:\n",
    "        for sub in [\"250M\", \"100M\", \"25M\", \"5M\"]:\n",
    "            #print(dict_rename[key].replace(\"50M\".lower(), sub.lower()))\n",
    "            dict_rename[key.replace(\"50M\", sub)] = dict_rename[key].replace(\"50M\".lower(), sub.lower())\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#dict_rename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.rename(columns=dict_rename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['var.chrom', 'var.pos_hg19', 'var.allele1', 'var.allele2', 'var.rsid',\n",
       "       'obs.estimate', 'obs.label', 'var.isused',\n",
       "       'pred.gkmsvm.encsr000emt.varscore.deltasvm',\n",
       "       'pred.enformer.encsr000emt.varscore.global_sad',\n",
       "       'pred.enformer.encsr000emt.varscore.global_sar',\n",
       "       'pred.chrombpnet.encsr637xsc.varscore.ips',\n",
       "       'pred.chrombpnet.encsr637xsc.varscore.pval',\n",
       "       'pred.chrombpnet.encsr637xsc.varscore.ies',\n",
       "       'pred.chrombpnet.encsr637xsc.varscore.jsd_x_aaq',\n",
       "       'pred.chrombpnet.encsr637xsc.varscore.logfc_x_aaq',\n",
       "       'ENCSR637XSC.jsd.mean', 'ENCSR637XSC.abs_logfc.mean',\n",
       "       'pred.chrombpnet.encsr637xsc.varscore.aaq',\n",
       "       'pred.chrombpnet.encsr637xsc.varscore.logfc',\n",
       "       'pred.chrombpnet.encsr637xsc.varscore.jsd',\n",
       "       'pred.chrombpnet.encsr000emt.varscore.ips',\n",
       "       'pred.chrombpnet.encsr000emt.varscore.pval',\n",
       "       'pred.chrombpnet.encsr000emt.varscore.ies',\n",
       "       'pred.chrombpnet.encsr000emt.varscore.jsd_x_aaq',\n",
       "       'pred.chrombpnet.encsr000emt.varscore.logfc_x_aaq',\n",
       "       'ENCSR000EMT.jsd.mean', 'ENCSR000EMT.abs_logfc.mean',\n",
       "       'pred.chrombpnet.encsr000emt.varscore.aaq',\n",
       "       'pred.chrombpnet.encsr000emt.varscore.logfc',\n",
       "       'pred.chrombpnet.encsr000emt.varscore.jsd',\n",
       "       'pred.chrombpnet.encsr637xsc_50m.varscore.ips',\n",
       "       'pred.chrombpnet.encsr637xsc_50m.varscore.pval',\n",
       "       'pred.chrombpnet.encsr637xsc_50m.varscore.ies',\n",
       "       'pred.chrombpnet.encsr637xsc_50m.varscore.jsd_x_aaq',\n",
       "       'pred.chrombpnet.encsr637xsc_50m.varscore.logfc_x_aaq',\n",
       "       'ENCSR637XSC.sub.50M.jsd.mean', 'ENCSR637XSC.sub.50M.abs_logfc.mean',\n",
       "       'pred.chrombpnet.encsr637xsc_50m.varscore.aaq',\n",
       "       'pred.chrombpnet.encsr637xsc_50m.varscore.logfc',\n",
       "       'pred.chrombpnet.encsr637xsc_50m.varscore.jsd',\n",
       "       'pred.chrombpnet.encsr637xsc_250m.varscore.ips',\n",
       "       'pred.chrombpnet.encsr637xsc_250m.varscore.pval',\n",
       "       'pred.chrombpnet.encsr637xsc_250m.varscore.logfc',\n",
       "       'pred.chrombpnet.encsr637xsc_250m.varscore.jsd',\n",
       "       'pred.chrombpnet.encsr637xsc_100m.varscore.ips',\n",
       "       'pred.chrombpnet.encsr637xsc_100m.varscore.pval',\n",
       "       'pred.chrombpnet.encsr637xsc_100m.varscore.logfc',\n",
       "       'pred.chrombpnet.encsr637xsc_100m.varscore.jsd',\n",
       "       'pred.chrombpnet.encsr637xsc_25m.varscore.ips',\n",
       "       'pred.chrombpnet.encsr637xsc_25m.varscore.pval',\n",
       "       'pred.chrombpnet.encsr637xsc_25m.varscore.logfc',\n",
       "       'pred.chrombpnet.encsr637xsc_25m.varscore.jsd',\n",
       "       'pred.chrombpnet.encsr637xsc_5m.varscore.ips',\n",
       "       'pred.chrombpnet.encsr637xsc_5m.varscore.pval',\n",
       "       'pred.chrombpnet.encsr637xsc_5m.varscore.logfc',\n",
       "       'pred.chrombpnet.encsr637xsc_5m.varscore.jsd',\n",
       "       'pred.enformer.encsr000emt.varscore.local_logfc'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "use_cols = ['var.chr', 'var.pos_hg19', 'var.allele1', 'var.allele2', 'var.rsid', 'obs.label', 'var.isused',\n",
    "       'obs.estimate', \n",
    "       'pred.gkmsvm.encsr000emt.varscore.deltasvm',\n",
    "       'pred.enformer.encsr000emt.varscore.global_sad',\n",
    "       'pred.enformer.encsr000emt.varscore.global_sar',\n",
    "       'pred.enformer.encsr000emt.varscore.local_logfc',\n",
    "       'pred.chrombpnet.encsr000emt.varscore.logfc',\n",
    "       'pred.chrombpnet.encsr000emt.varscore.jsd', \n",
    "       'pred.chrombpnet.encsr000emt.varscore.ips',\n",
    "       'pred.chrombpnet.encsr000emt.varscore.aaq',\n",
    "       'pred.chrombpnet.encsr000emt.varscore.ies',\n",
    "       'pred.chrombpnet.encsr000emt.varscore.jsd_x_aaq',\n",
    "       'pred.chrombpnet.encsr000emt.varscore.logfc_x_aaq', \n",
    "\n",
    "       'pred.chrombpnet.encsr637xsc.varscore.logfc',\n",
    "       'pred.chrombpnet.encsr637xsc.varscore.jsd', \n",
    "       'pred.chrombpnet.encsr637xsc.varscore.ips',\n",
    "       'pred.chrombpnet.encsr637xsc.varscore.aaq',\n",
    "       'pred.chrombpnet.encsr637xsc.varscore.ies',\n",
    "       'pred.chrombpnet.encsr637xsc.varscore.jsd_x_aaq',\n",
    "       'pred.chrombpnet.encsr637xsc.varscore.logfc_x_aaq', \n",
    "\n",
    "       'pred.chrombpnet.encsr637xsc_250m.varscore.logfc',\n",
    "       'pred.chrombpnet.encsr637xsc_250m.varscore.jsd', \n",
    "       'pred.chrombpnet.encsr637xsc_250m.varscore.ips',\n",
    "\n",
    "\n",
    "       'pred.chrombpnet.encsr637xsc_100m.varscore.logfc',\n",
    "       'pred.chrombpnet.encsr637xsc_100m.varscore.jsd', \n",
    "       'pred.chrombpnet.encsr637xsc_100m.varscore.ips',\n",
    "\n",
    "       'pred.chrombpnet.encsr637xsc_50m.varscore.logfc',\n",
    "       'pred.chrombpnet.encsr637xsc_50m.varscore.jsd', \n",
    "       'pred.chrombpnet.encsr637xsc_50m.varscore.ips',\n",
    "       'pred.chrombpnet.encsr637xsc_50m.varscore.aaq',\n",
    "       'pred.chrombpnet.encsr637xsc_50m.varscore.ies',\n",
    "       'pred.chrombpnet.encsr637xsc_50m.varscore.jsd_x_aaq',\n",
    "       'pred.chrombpnet.encsr637xsc_50m.varscore.logfc_x_aaq', \n",
    "\n",
    "       'pred.chrombpnet.encsr637xsc_25m.varscore.logfc',\n",
    "       'pred.chrombpnet.encsr637xsc_25m.varscore.jsd', \n",
    "       'pred.chrombpnet.encsr637xsc_25m.varscore.ips',\n",
    "\n",
    "       'pred.chrombpnet.encsr637xsc_5m.varscore.logfc',\n",
    "       'pred.chrombpnet.encsr637xsc_5m.varscore.jsd', \n",
    "       'pred.chrombpnet.encsr637xsc_5m.varscore.ips'  ]\n",
    "\n",
    "\n",
    " \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataf = data[use_cols]\n",
    "\n",
    "#data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var.chrom</th>\n",
       "      <th>var.pos_hg19</th>\n",
       "      <th>var.allele1</th>\n",
       "      <th>var.allele2</th>\n",
       "      <th>var.rsid</th>\n",
       "      <th>obs.label</th>\n",
       "      <th>var.isused</th>\n",
       "      <th>obs.estimate</th>\n",
       "      <th>pred.gkmsvm.encsr000emt.varscore.deltasvm</th>\n",
       "      <th>pred.enformer.encsr000emt.varscore.global_sad</th>\n",
       "      <th>...</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_50m.varscore.aaq</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_50m.varscore.ies</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_50m.varscore.jsd_x_aaq</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_50m.varscore.logfc_x_aaq</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_25m.varscore.logfc</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_25m.varscore.jsd</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_25m.varscore.ips</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_5m.varscore.logfc</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_5m.varscore.jsd</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_5m.varscore.ips</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>856583</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs141671872</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>2.653531</td>\n",
       "      <td>0.000342</td>\n",
       "      <td>...</td>\n",
       "      <td>0.248566</td>\n",
       "      <td>0.044994</td>\n",
       "      <td>0.021877</td>\n",
       "      <td>0.127298</td>\n",
       "      <td>0.291935</td>\n",
       "      <td>0.067401</td>\n",
       "      <td>0.004840</td>\n",
       "      <td>0.065098</td>\n",
       "      <td>0.049902</td>\n",
       "      <td>0.001360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>911595</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs7417106</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.6285</td>\n",
       "      <td>2.821422</td>\n",
       "      <td>-0.001241</td>\n",
       "      <td>...</td>\n",
       "      <td>0.726685</td>\n",
       "      <td>0.002535</td>\n",
       "      <td>0.027762</td>\n",
       "      <td>0.056287</td>\n",
       "      <td>0.093192</td>\n",
       "      <td>0.036053</td>\n",
       "      <td>0.002239</td>\n",
       "      <td>0.045314</td>\n",
       "      <td>0.034489</td>\n",
       "      <td>0.000898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1186502</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>rs6603785</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.7363</td>\n",
       "      <td>-7.167236</td>\n",
       "      <td>-0.006530</td>\n",
       "      <td>...</td>\n",
       "      <td>0.696101</td>\n",
       "      <td>0.021468</td>\n",
       "      <td>0.029658</td>\n",
       "      <td>0.349964</td>\n",
       "      <td>-0.402188</td>\n",
       "      <td>-0.038791</td>\n",
       "      <td>0.008839</td>\n",
       "      <td>-0.163023</td>\n",
       "      <td>-0.039080</td>\n",
       "      <td>0.002210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1227412</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs112571439</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.9236</td>\n",
       "      <td>-3.957382</td>\n",
       "      <td>-0.001878</td>\n",
       "      <td>...</td>\n",
       "      <td>0.521835</td>\n",
       "      <td>0.265543</td>\n",
       "      <td>0.137196</td>\n",
       "      <td>0.518532</td>\n",
       "      <td>-0.622704</td>\n",
       "      <td>-0.222396</td>\n",
       "      <td>0.086945</td>\n",
       "      <td>-0.259064</td>\n",
       "      <td>-0.172960</td>\n",
       "      <td>0.027583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1590575</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs9661285</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>-1.0890</td>\n",
       "      <td>-4.785596</td>\n",
       "      <td>-0.008736</td>\n",
       "      <td>...</td>\n",
       "      <td>0.988211</td>\n",
       "      <td>0.006033</td>\n",
       "      <td>0.053680</td>\n",
       "      <td>0.103350</td>\n",
       "      <td>-0.061906</td>\n",
       "      <td>-0.050589</td>\n",
       "      <td>0.003275</td>\n",
       "      <td>-0.032300</td>\n",
       "      <td>-0.023135</td>\n",
       "      <td>0.000802</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  var.chrom  var.pos_hg19 var.allele1 var.allele2     var.rsid  obs.label  \\\n",
       "0      chr1        856583           A           G  rs141671872          1   \n",
       "1      chr1        911595           G           A    rs7417106          1   \n",
       "2      chr1       1186502           T           A    rs6603785          1   \n",
       "3      chr1       1227412           A           G  rs112571439          1   \n",
       "4      chr1       1590575           A           G    rs9661285          1   \n",
       "\n",
       "   var.isused  obs.estimate  pred.gkmsvm.encsr000emt.varscore.deltasvm  \\\n",
       "0        True        0.9301                                   2.653531   \n",
       "1        True       -0.6285                                   2.821422   \n",
       "2        True       -0.7363                                  -7.167236   \n",
       "3        True       -0.9236                                  -3.957382   \n",
       "4        True       -1.0890                                  -4.785596   \n",
       "\n",
       "   pred.enformer.encsr000emt.varscore.global_sad  ...  \\\n",
       "0                                       0.000342  ...   \n",
       "1                                      -0.001241  ...   \n",
       "2                                      -0.006530  ...   \n",
       "3                                      -0.001878  ...   \n",
       "4                                      -0.008736  ...   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_50m.varscore.aaq  \\\n",
       "0                                      0.248566   \n",
       "1                                      0.726685   \n",
       "2                                      0.696101   \n",
       "3                                      0.521835   \n",
       "4                                      0.988211   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_50m.varscore.ies  \\\n",
       "0                                      0.044994   \n",
       "1                                      0.002535   \n",
       "2                                      0.021468   \n",
       "3                                      0.265543   \n",
       "4                                      0.006033   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_50m.varscore.jsd_x_aaq  \\\n",
       "0                                           0.021877    \n",
       "1                                           0.027762    \n",
       "2                                           0.029658    \n",
       "3                                           0.137196    \n",
       "4                                           0.053680    \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_50m.varscore.logfc_x_aaq  \\\n",
       "0                                           0.127298      \n",
       "1                                           0.056287      \n",
       "2                                           0.349964      \n",
       "3                                           0.518532      \n",
       "4                                           0.103350      \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_25m.varscore.logfc  \\\n",
       "0                                        0.291935   \n",
       "1                                        0.093192   \n",
       "2                                       -0.402188   \n",
       "3                                       -0.622704   \n",
       "4                                       -0.061906   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_25m.varscore.jsd  \\\n",
       "0                                      0.067401   \n",
       "1                                      0.036053   \n",
       "2                                     -0.038791   \n",
       "3                                     -0.222396   \n",
       "4                                     -0.050589   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_25m.varscore.ips  \\\n",
       "0                                      0.004840   \n",
       "1                                      0.002239   \n",
       "2                                      0.008839   \n",
       "3                                      0.086945   \n",
       "4                                      0.003275   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_5m.varscore.logfc  \\\n",
       "0                                       0.065098   \n",
       "1                                       0.045314   \n",
       "2                                      -0.163023   \n",
       "3                                      -0.259064   \n",
       "4                                      -0.032300   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_5m.varscore.jsd  \\\n",
       "0                                     0.049902   \n",
       "1                                     0.034489   \n",
       "2                                    -0.039080   \n",
       "3                                    -0.172960   \n",
       "4                                    -0.023135   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_5m.varscore.ips  \n",
       "0                                     0.001360  \n",
       "1                                     0.000898  \n",
       "2                                     0.002210  \n",
       "3                                     0.027583  \n",
       "4                                     0.000802  \n",
       "\n",
       "[5 rows x 45 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataf_final = dataf[dataf[\"var.isused\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(27373, 45)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataf_final.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataf_final = dataf_final.drop([\"var.isused\"], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var.chrom</th>\n",
       "      <th>var.pos_hg19</th>\n",
       "      <th>var.allele1</th>\n",
       "      <th>var.allele2</th>\n",
       "      <th>var.rsid</th>\n",
       "      <th>obs.label</th>\n",
       "      <th>obs.estimate</th>\n",
       "      <th>pred.gkmsvm.encsr000emt.varscore.deltasvm</th>\n",
       "      <th>pred.enformer.encsr000emt.varscore.global_sad</th>\n",
       "      <th>pred.enformer.encsr000emt.varscore.global_sar</th>\n",
       "      <th>...</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_50m.varscore.aaq</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_50m.varscore.ies</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_50m.varscore.jsd_x_aaq</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_50m.varscore.logfc_x_aaq</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_25m.varscore.logfc</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_25m.varscore.jsd</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_25m.varscore.ips</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_5m.varscore.logfc</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_5m.varscore.jsd</th>\n",
       "      <th>pred.chrombpnet.encsr637xsc_5m.varscore.ips</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>856583</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs141671872</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9301</td>\n",
       "      <td>2.653531</td>\n",
       "      <td>0.000342</td>\n",
       "      <td>0.000450</td>\n",
       "      <td>...</td>\n",
       "      <td>0.248566</td>\n",
       "      <td>0.044994</td>\n",
       "      <td>0.021877</td>\n",
       "      <td>0.127298</td>\n",
       "      <td>0.291935</td>\n",
       "      <td>0.067401</td>\n",
       "      <td>0.004840</td>\n",
       "      <td>0.065098</td>\n",
       "      <td>0.049902</td>\n",
       "      <td>0.001360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>911595</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs7417106</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.6285</td>\n",
       "      <td>2.821422</td>\n",
       "      <td>-0.001241</td>\n",
       "      <td>-0.001526</td>\n",
       "      <td>...</td>\n",
       "      <td>0.726685</td>\n",
       "      <td>0.002535</td>\n",
       "      <td>0.027762</td>\n",
       "      <td>0.056287</td>\n",
       "      <td>0.093192</td>\n",
       "      <td>0.036053</td>\n",
       "      <td>0.002239</td>\n",
       "      <td>0.045314</td>\n",
       "      <td>0.034489</td>\n",
       "      <td>0.000898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1186502</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>rs6603785</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.7363</td>\n",
       "      <td>-7.167236</td>\n",
       "      <td>-0.006530</td>\n",
       "      <td>-0.007840</td>\n",
       "      <td>...</td>\n",
       "      <td>0.696101</td>\n",
       "      <td>0.021468</td>\n",
       "      <td>0.029658</td>\n",
       "      <td>0.349964</td>\n",
       "      <td>-0.402188</td>\n",
       "      <td>-0.038791</td>\n",
       "      <td>0.008839</td>\n",
       "      <td>-0.163023</td>\n",
       "      <td>-0.039080</td>\n",
       "      <td>0.002210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1227412</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs112571439</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.9236</td>\n",
       "      <td>-3.957382</td>\n",
       "      <td>-0.001878</td>\n",
       "      <td>-0.002290</td>\n",
       "      <td>...</td>\n",
       "      <td>0.521835</td>\n",
       "      <td>0.265543</td>\n",
       "      <td>0.137196</td>\n",
       "      <td>0.518532</td>\n",
       "      <td>-0.622704</td>\n",
       "      <td>-0.222396</td>\n",
       "      <td>0.086945</td>\n",
       "      <td>-0.259064</td>\n",
       "      <td>-0.172960</td>\n",
       "      <td>0.027583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1590575</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>rs9661285</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.0890</td>\n",
       "      <td>-4.785596</td>\n",
       "      <td>-0.008736</td>\n",
       "      <td>-0.010230</td>\n",
       "      <td>...</td>\n",
       "      <td>0.988211</td>\n",
       "      <td>0.006033</td>\n",
       "      <td>0.053680</td>\n",
       "      <td>0.103350</td>\n",
       "      <td>-0.061906</td>\n",
       "      <td>-0.050589</td>\n",
       "      <td>0.003275</td>\n",
       "      <td>-0.032300</td>\n",
       "      <td>-0.023135</td>\n",
       "      <td>0.000802</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  var.chrom  var.pos_hg19 var.allele1 var.allele2     var.rsid  obs.label  \\\n",
       "0      chr1        856583           A           G  rs141671872          1   \n",
       "1      chr1        911595           G           A    rs7417106          1   \n",
       "2      chr1       1186502           T           A    rs6603785          1   \n",
       "3      chr1       1227412           A           G  rs112571439          1   \n",
       "4      chr1       1590575           A           G    rs9661285          1   \n",
       "\n",
       "   obs.estimate  pred.gkmsvm.encsr000emt.varscore.deltasvm  \\\n",
       "0        0.9301                                   2.653531   \n",
       "1       -0.6285                                   2.821422   \n",
       "2       -0.7363                                  -7.167236   \n",
       "3       -0.9236                                  -3.957382   \n",
       "4       -1.0890                                  -4.785596   \n",
       "\n",
       "   pred.enformer.encsr000emt.varscore.global_sad  \\\n",
       "0                                       0.000342   \n",
       "1                                      -0.001241   \n",
       "2                                      -0.006530   \n",
       "3                                      -0.001878   \n",
       "4                                      -0.008736   \n",
       "\n",
       "   pred.enformer.encsr000emt.varscore.global_sar  ...  \\\n",
       "0                                       0.000450  ...   \n",
       "1                                      -0.001526  ...   \n",
       "2                                      -0.007840  ...   \n",
       "3                                      -0.002290  ...   \n",
       "4                                      -0.010230  ...   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_50m.varscore.aaq  \\\n",
       "0                                      0.248566   \n",
       "1                                      0.726685   \n",
       "2                                      0.696101   \n",
       "3                                      0.521835   \n",
       "4                                      0.988211   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_50m.varscore.ies  \\\n",
       "0                                      0.044994   \n",
       "1                                      0.002535   \n",
       "2                                      0.021468   \n",
       "3                                      0.265543   \n",
       "4                                      0.006033   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_50m.varscore.jsd_x_aaq  \\\n",
       "0                                           0.021877    \n",
       "1                                           0.027762    \n",
       "2                                           0.029658    \n",
       "3                                           0.137196    \n",
       "4                                           0.053680    \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_50m.varscore.logfc_x_aaq  \\\n",
       "0                                           0.127298      \n",
       "1                                           0.056287      \n",
       "2                                           0.349964      \n",
       "3                                           0.518532      \n",
       "4                                           0.103350      \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_25m.varscore.logfc  \\\n",
       "0                                        0.291935   \n",
       "1                                        0.093192   \n",
       "2                                       -0.402188   \n",
       "3                                       -0.622704   \n",
       "4                                       -0.061906   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_25m.varscore.jsd  \\\n",
       "0                                      0.067401   \n",
       "1                                      0.036053   \n",
       "2                                     -0.038791   \n",
       "3                                     -0.222396   \n",
       "4                                     -0.050589   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_25m.varscore.ips  \\\n",
       "0                                      0.004840   \n",
       "1                                      0.002239   \n",
       "2                                      0.008839   \n",
       "3                                      0.086945   \n",
       "4                                      0.003275   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_5m.varscore.logfc  \\\n",
       "0                                       0.065098   \n",
       "1                                       0.045314   \n",
       "2                                      -0.163023   \n",
       "3                                      -0.259064   \n",
       "4                                      -0.032300   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_5m.varscore.jsd  \\\n",
       "0                                     0.049902   \n",
       "1                                     0.034489   \n",
       "2                                    -0.039080   \n",
       "3                                    -0.172960   \n",
       "4                                    -0.023135   \n",
       "\n",
       "   pred.chrombpnet.encsr637xsc_5m.varscore.ips  \n",
       "0                                     0.001360  \n",
       "1                                     0.000898  \n",
       "2                                     0.002210  \n",
       "3                                     0.027583  \n",
       "4                                     0.000802  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataf_final.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert (sum(dataf.columns.isin(use_cols))==dataf.shape[1])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataf.to_csv(\"full_variants/dsqtls.yoruba.lcls.benchmarking.all.tsv\", sep='\\t', header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataf_final.to_csv(\"filtered_variants/dsqtls.yoruba.lcls.benchmarking.chrombpnet.tsv\", sep='\\t', header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "readme_data = {'var.chrom' : \"Chromosome of the variant (GRCh37)\", \n",
    "        'var.pos_hg19': \"Position of the variant (GRCh37, 1-based)\",\n",
    "        'var.allele1': \"Allele 1 for the variant\", \n",
    "        'var.allele2': \"Allele 2 for the variant\",        \n",
    "        'var.rsid': \"dbSNP rsid identifier\", \n",
    "       'obs.label': \"True if variant has significant observed effect size\", \n",
    "       'var.isused': \"True if variant is used in final ChromBPNet benchmarking\",\n",
    "       'obs.estimate': \"Observed effect size\", \n",
    "       'pred.gkmsvm.encsr000emt.varscore.deltasvm' : \"deltasvm predictions in encsr000emt\",\n",
    "       'pred.enformer.encsr000emt.varscore.global_sad': \"Enformer SAD predictions (downloaded) in encsr000emt\",\n",
    "       'pred.enformer.encsr000emt.varscore.global_sar': \"Enformer SAR predictions (downloaded) in encsr000emt\",\n",
    "       'pred.enformer.encsr000emt.varscore.local_logfc': \"Enformer logFC (local) predictions (re-computed) in encsr000emt\",\n",
    "       'pred.chrombpnet.xxx.varscore.logfc': \"ChromBPNet logFC predictions in encid xxx\",\n",
    "       'pred.chrombpnet.xxx.varscore.jsd':  \"ChromBPNet JSD predictions in encid xxx\", \n",
    "       'pred.chrombpnet.xxx.varscore.ips': \"ChromBPNet IPS predictions in encid xxx\",\n",
    "       'pred.chrombpnet.xxx.varscore.aaq': \"ChromBPNet AAQ predictions in encid xxx\",\n",
    "       'pred.chrombpnet.xxx.varscore.ies': \"ChromBPNet IES predictions in encid xxx\",\n",
    "       'pred.chrombpnet.xxx.varscore.jsd_x_aaq': \"ChromBPNet JSD x AAQ predictions in encid xxx\",\n",
    "       'pred.chrombpnet.xxx.varscore.logfc_x_aaq': \"ChromBPNet logFC x AAQ predictions in encid xxx\"}\n",
    "\n",
    "# encsr637xsc_5m\n",
    "# encsr637xsc_25m\n",
    "# encsr637xsc_50m\n",
    "# encsr637xsc_100m\n",
    "# encsr637xsc_250m\n",
    "# encsr637xsc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "var.chrom\n",
      "var.pos_hg19\n",
      "var.allele1\n",
      "var.allele2\n",
      "var.rsid\n",
      "obs.label\n",
      "var.isused\n",
      "obs.estimate\n",
      "pred.gkmsvm.encsr000emt.varscore.deltasvm\n",
      "pred.enformer.encsr000emt.varscore.global_sad\n",
      "pred.enformer.encsr000emt.varscore.global_sar\n",
      "pred.enformer.encsr000emt.varscore.local_logfc\n",
      "pred.chrombpnet.xxx.varscore.logfc\n",
      "pred.chrombpnet.xxx.varscore.jsd\n",
      "pred.chrombpnet.xxx.varscore.ips\n",
      "pred.chrombpnet.xxx.varscore.aaq\n",
      "pred.chrombpnet.xxx.varscore.ies\n",
      "pred.chrombpnet.xxx.varscore.jsd_x_aaq\n",
      "pred.chrombpnet.xxx.varscore.logfc_x_aaq\n"
     ]
    }
   ],
   "source": [
    "f = open(\"full_variants/dsqtls.yoruba.lcls.benchmarking.all.README\", \"w\")\n",
    "f.write(\"# TSV Field Header Descriptions \\n \\n\")\n",
    "for val in readme_data:\n",
    "    print(val)\n",
    "    f.write(val+\" : \"+readme_data[val]+\"\\n\")\n",
    "f.write(\"\\n\")\n",
    "f.write(\"# Experiment-Specific Headers and IDs  \\n\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
